{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a0d93c-1359-4107-807f-f230b3863708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89dac285-5db9-4261-b4e4-38e0f84db137",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_read = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens.GRCh38.112.chr.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "#tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f9135a3-08a2-46ed-9c66-2622790d4290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3516515   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3516516   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3516517   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3516518   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3516519   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                        Identifiers  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3516515      -       .                                 logic_name=eponine  \n",
       "3516516      +       .                                 logic_name=eponine  \n",
       "3516517      -       .                                 logic_name=eponine  \n",
       "3516518      +       .                                 logic_name=eponine  \n",
       "3516519      -       .                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Signal', 7: 'Unknown', 8: 'Identifiers'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64551187-3a83-4376-a7db-f0cc87b190b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c35e5d62-69a8-44a0-be89-734bae0489bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=chromosome:1;Alias=CM000663.2,chr1,NC_000001.11'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Identifiers\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8d7245-d6be-4015-bce8-07cb7093450f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "#genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "#CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "#UTR3 = tanscripts_df.loc[tanscripts_df['Type'] == \"three_prime_UTR\"]\n",
    "#UTR5 = tanscripts_df.loc[tanscripts_df['Type'] == \"five_prime_UTR\"]\n",
    "#exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49bbd252-0fed-472e-a11c-f5eba8d3f97c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_overlap = UTR5.loc[UTR5[\"Identifiers\"] == UTR5[\"Identifiers\"][3516301]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cd7ce5a1-dd7c-445d-8388-e423544af64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_dict_toDF(dictionary):\n",
    "    transcripts_matrix = []\n",
    "    concat_list = []\n",
    "    for value in dictionary.values():\n",
    "        concat_list.append(value[0])\n",
    "    return pd.concat(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86fb3cf3-3746-4a7b-ae73-8dafdbf4be72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_NoOverlaps(element_type = \"CDS\"):\n",
    "    element_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    elements_dict = {}\n",
    "    #elements_dict.update({\"header\": element_df.columns.tolist()})\n",
    "    count = 0\n",
    "    for row in element_df.index:\n",
    "        identifiers = element_df[\"Identifiers\"][row]\n",
    "        transcript_id = identifiers.split(\";\")[0]\n",
    "        transcript_length = element_df[\"Last_index\"][row] - element_df[\"First_index\"][row]\n",
    "        if transcript_id not in elements_dict.keys():\n",
    "        #if elements_dict.get(transcript_id) == None:\n",
    "            #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "            elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        else:\n",
    "            #print(\"resolving overlap...\")\n",
    "            if transcript_length > elements_dict[transcript_id][1]:\n",
    "                #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "                elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        count+=1\n",
    "    return transcripts_dict_toDF(elements_dict)\n",
    "    #transcripts_matrix = []\n",
    "    #for value in elements_dict.values():\n",
    "    #    transcripts_matrix.append(value[0])\n",
    "    #return elements_dict\n",
    "    #return pd.DataFrame(elements_dict.values())\n",
    "    #return pd.DataFrame(transcripts_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a1b2b71-b9c0-4267-9e61-2a142000233f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fivePrime_UTR_dict = transcripts_NoOverlaps(\"five_prime_UTR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7a064b-470e-412d-b993-49f90c35aa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for value in fivePrime_UTR_dict.values():\n",
    "#    print(type(value[0]))\n",
    "#    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0006eaa1-0cac-4ed6-bb12-05c631af2605",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df = transcripts_dict_toDF(fivePrime_UTR_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0554414-1b4b-4fd1-8fd6-9d1269d206f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171e9519-2e5a-4085-82e0-8b2b12b53fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df[\"Identifiers\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "867e6833-f85a-473a-add0-e59b583c5f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_matrix(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    The matrix representing all intances of a sequence type available from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        #print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            chr_array[int(chr_sequence[\"First_index\"][row]) : int(chr_sequence[\"Last_index\"][row])] += 1\n",
    "                \n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0053bc01-6bdd-46e1-9592-41e0f5a2599b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity(cytobands_df, region_matrix, col_name=\"Element_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    densities = []\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            dr_length = np.count_nonzero(cytoband_array)\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            density = dr_length/cytoband_length\n",
    "            densities.append(round(density, 10))\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9e6e7d8a-9c5d-4c16-95ba-720845fc909c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity(cytobands_df, region_matrix, exons_matrix, col_name1=\"exon_density\", col_name2=\"intron_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):\n",
    "        #print(\"Calculating densities in chromosome \", cytobands_df[\"Chromossome\"].unique()[chrom], \" ...\")\n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            ############################ Resolver problema, o conjunto dos exons contem todo o conjunto dos CDSs ###############################\n",
    "            region_cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            exon_cytoband_array = exons_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            #print(\"cytoband: \", chr_cytobands[\"Citoband\"][row])\n",
    "            #print(\"array_regiao: \", region_cytoband_array)\n",
    "            #print(\"exon_regiao: \", exon_cytoband_array)\n",
    "            region_indexes = np.where(region_cytoband_array != 0)[0].tolist()\n",
    "            exon_indexes = np.where(exon_cytoband_array != 0)[0].tolist()\n",
    "            #print(\"number of elements of region: \", len(region_indexes))\n",
    "            #print(\"number of exons in region: \", len(np.intersect1d(region_indexes, exon_indexes)))\n",
    "            #print(\"exons in CDS region: \", np.intersect1d(region_indexes, exon_indexes))\n",
    "            #print(\"introns in CDS region: \", np.setdiff1d(region_indexes, exon_indexes))\n",
    "            exon_length = len(np.intersect1d(region_indexes, exon_indexes))    #indices que estam na região e são exões\n",
    "            intron_length = len(np.setdiff1d(region_indexes, exon_indexes))      #indices que estam na região mas não são exões\n",
    "            #for region_element in region_indexes:\n",
    "            #    if region_element in exon_indexes:\n",
    "            #        exon_length+=1\n",
    "            #    elif region_element not in exon_indexes:\n",
    "            #        intron_length+=1\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            exon_density = exon_length/cytoband_length\n",
    "            intron_density = intron_length/cytoband_length\n",
    "            exon_densities.append(round(exon_density, 10))\n",
    "            intron_densities.append(round(intron_density, 10))   #max 6\n",
    "            \n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name1, exon_densities, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name2, intron_densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93fd572c-e059-4592-8147-cbeb4f16de06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2 3 7 8]\n",
      "[1 6]\n"
     ]
    }
   ],
   "source": [
    "array1 = [1, 2, 3, 6, 7, 8]\n",
    "array2 = [2, 3, 4, 7, 8, 9]\n",
    "print(np.intersect1d(array1, array2))\n",
    "print(np.setdiff1d(array1, array2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "84e048d0-2f72-453e-bdb7-d9007893a020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    df_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        df_cols += col + \"\\t\"\n",
    "    df_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(df_cols))\n",
    "    for row in df.values:\n",
    "        df_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            df_row += str(col) + \"\\t\"\n",
    "        df_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(df_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1e8cdd-042e-4071-9737-69169c01ee2d",
   "metadata": {},
   "source": [
    "# Code running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fd457d08-3d9c-4ad7-b34a-e4cdd5aa389d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  \n",
       "0              0.087516   0.006384   0.084476     0.326328  \n",
       "1              0.084077   0.013902   0.079223     0.281556  \n",
       "2              0.056236   0.012909   0.051421     0.275410  \n",
       "3              0.049082   0.013229   0.042217     0.404844  \n",
       "4              0.055881   0.011072   0.048928     0.404047  \n",
       "..                  ...        ...        ...          ...  \n",
       "858            0.036956   0.014211   0.029356     0.015158  \n",
       "859            0.061662   0.015902   0.052790     0.012965  \n",
       "860            0.082234   0.005789   0.078347     0.006441  \n",
       "861            0.026278   0.009164   0.020721     0.001822  \n",
       "862            0.005530   0.001643   0.004251     0.000594  \n",
       "\n",
       "[863 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoband_read = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoband_write = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr_dg_MANESelect.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoband_read, sep=\"\\t\")#, header=None))\n",
    "#cytobands_df2 = pd.read_csv(cytoband_write, sep=\"\\t\")\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5377fb6f-0db3-40d0-9c18-1632ee187249",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd5b333-668d-4f62-94f9-a253756f0d38",
   "metadata": {},
   "source": [
    "## Tirar as linhas com a tag \"MANE_Plus_Critical\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8885c4c9-6885-4df0-b83a-89817e3c6468",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mask1 = (\"MANE_Plus_Critical\" in tanscripts_df['Identifiers']) == 1)# & (gvs_chr20_df['ALT'].str.len() == 1)\n",
    "#tanscripts_df_ManeSelect = tanscripts_df[~tanscripts_df['Identifiers'].str.contains(\"MANE_Plus_Critical\")]\n",
    "MPC_rows = tanscripts_df['Identifiers'].str.contains(\"MANE_Plus_Clinical\", na=False)\n",
    "tanscripts_df_ManeSelect = tanscripts_df[~MPC_rows]\n",
    "#tanscripts_df_ManeSelect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03477a94-ef8a-4684-be34-7a0ca6b91304",
   "metadata": {},
   "source": [
    "#### Foram retiradas 64 linhas com a tag \"MANE_Plus_Clinical\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "083cfb80-3439-436b-abfc-8426f5ed1dc5",
   "metadata": {},
   "source": [
    "### Criação de variáveis globais necessárias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0fa9158f-d079-4c6e-bedd-6e8072a100ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Para o dataset sem filtragem #################################################################\n",
    "chromosomes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "exon_matrix = element_matrix(chromosomes, exons)\n",
    "\n",
    "############################ Para o dataset com filtragem pela tag ########################################################\n",
    "#chromosomes_ManeSelect = tanscripts_df_ManeSelect.loc[tanscripts_df_ManeSelect['Type'] == \"chromosome\"]\n",
    "#exons_ManeSelect = tanscripts_df_ManeSelect.loc[tanscripts_df_ManeSelect['Type'] == \"exon\"]\n",
    "#exon_matrix_ManeSelect = element_matrix(chromosomes_ManeSelect, exons_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "16adb942-48c6-4bda-8d19-2a5425552610",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Para o dataset com filtragem de transcritos ###################################################\n",
    "#exons_noOverlaps_df = transcripts_NoOverlaps(\"exon\")\n",
    "#exon_matrix_noOvrlp = element_matrix(chromosomes, exons_noOverlaps_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "630588ea-ca59-42bb-8713-6f74e47dd2ad",
   "metadata": {},
   "source": [
    "## Sem filtragem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3c1d4022-9ac9-4907-97ad-6ee7b43491af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_transcriptDens(cytobands_df, element_type = \"CDS\", tanscripts_df = tanscripts_df, chromosomes_df = chromosomes):\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    sub_trascripts_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    print(\"building matriz...\")\n",
    "    transcripts_matrix = element_matrix(chromosomes_df, sub_trascripts_df)\n",
    "    colname = element_type+\"s_density\"\n",
    "    print(\"doing calculations...\")\n",
    "    calc_elementDensity(cytobands_df, transcripts_matrix, colname)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "07516778-0141-486b-814d-ad58bdb085ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_Trcpts_IExons(cytobands_df, element_type = \"CDS\", tanscripts_df = tanscripts_df, chromosomes_df = chromosomes, exon_matrix = exon_matrix):\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    sub_trascripts_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "    print(\"building matriz...\")\n",
    "    transcripts_matrix = element_matrix(chromosomes_df, sub_trascripts_df)\n",
    "    exon_matrix = element_matrix(chromosomes_df, exons)\n",
    "    print(\"doing calculations...\")\n",
    "    colname1 = element_type+\"Exon_density\"\n",
    "    colname2 = element_type+\"Intron_density\"\n",
    "    calc_ExonIntronDensity(cytobands_df, transcripts_matrix, exon_matrix, colname1, colname2)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c4dbff3-8457-4e99-b1a6-aca386e290e2",
   "metadata": {},
   "source": [
    "## Com filtragem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b6fb2715-301c-4988-acbf-e2f877c85b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_transcriptDens_noOverlps(cytobands_df, element_type = \"CDS\", chromosomes_df = chromosomes):\n",
    "    chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_noOverlaps_df = transcripts_NoOverlaps(element_type)\n",
    "    print(\"building matriz...\")\n",
    "    transcript_matrix = element_matrix(chromosomes_df, element_noOverlaps_df)\n",
    "    colname = element_type+\"s_density\"\n",
    "    print(\"doing calculations...\")\n",
    "    calc_elementDensity(cytobands_df, transcript_matrix, colname)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68b2db56-016e-4314-a28d-405156aa373a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_Trcpts_IExons_noOverlps(cytobands_df, element_type = \"CDS\", chromosomes_df = chromosomes, exon_matrix_noOvrlp = exon_matrix_noOvrlp):\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_noOverlaps_df = transcripts_NoOverlaps(element_type)\n",
    "    #exons_dict = transcripts_NoOverlaps(\"exon\")\n",
    "    #exons_noOverlaps_df = transcripts_dict_toDF(exons_dict)\n",
    "    print(\"a fazer a matriz...\")\n",
    "    transcript_matrix = element_matrix(chromosomes_df, element_noOverlaps_df)\n",
    "    #exon_matrix = element_matrix(chromosmes, exons_noOverlaps_df)\n",
    "    colname1 = element_type+\"Intron_density\"\n",
    "    colname2 = element_type+\"Exon_density\"\n",
    "    print(\"a fazer os calculos...\")\n",
    "    calc_ExonIntronDensity(cytobands_df, transcript_matrix, exon_matrix_noOvrlp, colname1, colname2)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01879f35-00d3-4754-89a5-1b34a9447e01",
   "metadata": {},
   "source": [
    "## Com Nova filtragem é aplicar as funções sem filtragem no dataframe já filtrado: tanscripts_df_ManeSelect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9495e9-9702-4571-a78f-9d4835b82301",
   "metadata": {},
   "source": [
    "## CDS density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2e61635-55f7-4429-9bca-4478b22c7104",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n"
     ]
    }
   ],
   "source": [
    "#study_transcriptDens(cytobands_df, \"CDS\")\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"CDS\")\n",
    "study_transcriptDens(cytobands_df, \"CDS\", tanscripts_df_ManeSelect, chromosomes_ManeSelect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d690d7f-d2aa-453e-b132-1ac6e85dcf68",
   "metadata": {},
   "source": [
    "### CDSs' Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d2701f6-bff2-4b7e-ab92-d701c1bed997",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"CDS\")\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"CDS\")\n",
    "study_Trcpts_IExons(cytobands_df, \"CDS\", tanscripts_df_ManeSelect, chromosomes_ManeSelect, exon_matrix_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b6d07a5-655d-44d4-a2ad-8ea1b2f48ad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b149728d-101f-4b84-9fa8-a77b5686e219",
   "metadata": {},
   "source": [
    "## 5 Prime UTR density studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2778db1a-3cd0-46d2-961c-4a3b807334a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_transcriptDens(cytobands_df, \"five_prime_UTR\")\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"five_prime_UTR\")\n",
    "study_transcriptDens(cytobands_df, \"five_prime_UTR\", tanscripts_df_ManeSelect, chromosomes_ManeSelect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bd2ce3-76ac-481a-a4af-20733256d243",
   "metadata": {},
   "source": [
    "### 5 Prime UTR Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c4b5e9-f175-4756-9342-b832d4317a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"five_prime_UTR\")\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"five_prime_UTR\")\n",
    "study_Trcpts_IExons(cytobands_df, \"five_prime_UTR\", tanscripts_df_ManeSelect, chromosomes_ManeSelect, exon_matrix_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "184ef8f3-92ec-4ed3-9b38-8ed4b3d0fb50",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51f8d5e-d87f-47bc-89ef-60478ed00610",
   "metadata": {},
   "source": [
    "## Prime 3 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b70b50a-9eae-4fa6-932b-9ce572f7fd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_transcriptDens(cytobands_df, \"three_prime_UTR\")\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"three_prime_UTR\")\n",
    "study_transcriptDens(cytobands_df, \"three_prime_UTR\", tanscripts_df_ManeSelect, chromosomes_ManeSelect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d99a6f99-a946-4123-8f99-8984c9b011a0",
   "metadata": {},
   "source": [
    "## 3 Prime UTR Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6f4ede0-3475-415d-995e-5096c30268f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"three_prime_UTR\")\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"three_prime_UTR\")\n",
    "study_Trcpts_IExons(cytobands_df, \"three_prime_UTR\", tanscripts_df_ManeSelect, chromosomes_ManeSelect, exon_matrix_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383b40f1-7696-41e8-bb07-3dceca86bd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc99fd3-b968-4579-9ce9-c0ceec7f6925",
   "metadata": {},
   "source": [
    "## Exons density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e17263-abd9-4634-b815-4ef21afd7bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#study_transcriptDens(cytobands_df, \"exon\")\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"exon\")\n",
    "study_transcriptDens(cytobands_df, \"exon\", tanscripts_df_ManeSelect, chromosomes_ManeSelect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71d7563-10f1-4782-a714-2dc899e78b2d",
   "metadata": {},
   "source": [
    "## Resultados finais no dataset de cytobands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d902ad05-249f-4cd0-b698-a41e53d254b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e597745-bcfe-4359-be42-3b8d639cc3f6",
   "metadata": {},
   "source": [
    "## Resultados depois de filtrar apenas as região de maior tamanho de cada transcrito  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "520d1578-d4d3-4b78-9768-d098f23a045f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.016467</td>\n",
       "      <td>0.013392</td>\n",
       "      <td>0.003076</td>\n",
       "      <td>0.010664</td>\n",
       "      <td>0.005135</td>\n",
       "      <td>0.005529</td>\n",
       "      <td>0.026310</td>\n",
       "      <td>0.025853</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>0.133749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>0.004490</td>\n",
       "      <td>0.001576</td>\n",
       "      <td>0.003274</td>\n",
       "      <td>0.001761</td>\n",
       "      <td>0.001512</td>\n",
       "      <td>0.019776</td>\n",
       "      <td>0.019406</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>0.060012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.010266</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0.003362</td>\n",
       "      <td>0.005659</td>\n",
       "      <td>0.002073</td>\n",
       "      <td>0.003586</td>\n",
       "      <td>0.025848</td>\n",
       "      <td>0.025715</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.056062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.005996</td>\n",
       "      <td>0.004821</td>\n",
       "      <td>0.001175</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.001499</td>\n",
       "      <td>0.015875</td>\n",
       "      <td>0.015746</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.034280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.009799</td>\n",
       "      <td>0.006866</td>\n",
       "      <td>0.002934</td>\n",
       "      <td>0.004954</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.002777</td>\n",
       "      <td>0.029553</td>\n",
       "      <td>0.028516</td>\n",
       "      <td>0.001037</td>\n",
       "      <td>0.066018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000927</td>\n",
       "      <td>0.000344</td>\n",
       "      <td>0.000583</td>\n",
       "      <td>0.002534</td>\n",
       "      <td>0.002459</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.009953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>0.000802</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.000540</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000503</td>\n",
       "      <td>0.002733</td>\n",
       "      <td>0.002733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.001294</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.002016</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.016467   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.006066   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.010266   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.005996   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.009799   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.000917   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002263   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.001096   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.001369   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000050   \n",
       "\n",
       "     CDSIntron_density  CDSExon_density  five_prime_UTRs_density  \\\n",
       "0             0.013392         0.003076                 0.010664   \n",
       "1             0.004490         0.001576                 0.003274   \n",
       "2             0.006903         0.003362                 0.005659   \n",
       "3             0.004821         0.001175                 0.003569   \n",
       "4             0.006866         0.002934                 0.004954   \n",
       "..                 ...              ...                      ...   \n",
       "858           0.000627         0.000290                 0.000927   \n",
       "859           0.002263         0.000000                 0.000346   \n",
       "860           0.000802         0.000294                 0.000540   \n",
       "861           0.001294         0.000075                 0.000652   \n",
       "862           0.000043         0.000007                 0.000005   \n",
       "\n",
       "     five_prime_UTRIntron_density  five_prime_UTRExon_density  \\\n",
       "0                        0.005135                    0.005529   \n",
       "1                        0.001761                    0.001512   \n",
       "2                        0.002073                    0.003586   \n",
       "3                        0.002070                    0.001499   \n",
       "4                        0.002176                    0.002777   \n",
       "..                            ...                         ...   \n",
       "858                      0.000344                    0.000583   \n",
       "859                      0.000338                    0.000008   \n",
       "860                      0.000037                    0.000503   \n",
       "861                      0.000345                    0.000308   \n",
       "862                      0.000000                    0.000005   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRIntron_density  \\\n",
       "0                    0.026310                       0.025853   \n",
       "1                    0.019776                       0.019406   \n",
       "2                    0.025848                       0.025715   \n",
       "3                    0.015875                       0.015746   \n",
       "4                    0.029553                       0.028516   \n",
       "..                        ...                            ...   \n",
       "858                  0.002534                       0.002459   \n",
       "859                  0.000937                       0.000635   \n",
       "860                  0.002733                       0.002733   \n",
       "861                  0.002016                       0.001749   \n",
       "862                  0.000094                       0.000094   \n",
       "\n",
       "     three_prime_UTRExon_density  exons_density  \n",
       "0                       0.000456       0.133749  \n",
       "1                       0.000370       0.060012  \n",
       "2                       0.000133       0.056062  \n",
       "3                       0.000129       0.034280  \n",
       "4                       0.001037       0.066018  \n",
       "..                           ...            ...  \n",
       "858                     0.000075       0.009953  \n",
       "859                     0.000302       0.033020  \n",
       "860                     0.000000       0.021283  \n",
       "861                     0.000267       0.022278  \n",
       "862                     0.000000       0.000460  \n",
       "\n",
       "[863 rows x 20 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "510a3d3b-6634-4ee4-a067-2fa62b571fe3",
   "metadata": {},
   "source": [
    "## Resultados sem usar filtros, o conjunto dos exões cobre completamente as regiões estudadas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "238d1646-ee79-45ce-b69b-99bd9992f480",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "      <th>exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.078555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.082023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.092777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  five_prime_UTRs_density  \\\n",
       "0           0.037748                0.0                 0.012022   \n",
       "1           0.015817                0.0                 0.003561   \n",
       "2           0.024363                0.0                 0.006496   \n",
       "3           0.012765                0.0                 0.004600   \n",
       "4           0.029924                0.0                 0.005499   \n",
       "..               ...                ...                      ...   \n",
       "858         0.003879                0.0                 0.001027   \n",
       "859         0.002446                0.0                 0.000346   \n",
       "860         0.004806                0.0                 0.000675   \n",
       "861         0.002666                0.0                 0.000838   \n",
       "862         0.000130                0.0                 0.000006   \n",
       "\n",
       "     five_prime_UTRExon_density  five_prime_UTRIntron_density  \\\n",
       "0                      0.012022                           0.0   \n",
       "1                      0.003561                           0.0   \n",
       "2                      0.006496                           0.0   \n",
       "3                      0.004600                           0.0   \n",
       "4                      0.005499                           0.0   \n",
       "..                          ...                           ...   \n",
       "858                    0.001027                           0.0   \n",
       "859                    0.000346                           0.0   \n",
       "860                    0.000675                           0.0   \n",
       "861                    0.000838                           0.0   \n",
       "862                    0.000006                           0.0   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRExon_density  \\\n",
       "0                    0.032887                     0.032887   \n",
       "1                    0.022181                     0.022181   \n",
       "2                    0.033723                     0.033723   \n",
       "3                    0.017843                     0.017843   \n",
       "4                    0.035758                     0.035758   \n",
       "..                        ...                          ...   \n",
       "858                  0.002867                     0.002867   \n",
       "859                  0.001032                     0.001032   \n",
       "860                  0.003568                     0.003568   \n",
       "861                  0.002204                     0.002204   \n",
       "862                  0.000104                     0.000104   \n",
       "\n",
       "     three_prime_UTRIntron_density  exons_density  \n",
       "0                              0.0       0.170647  \n",
       "1                              0.0       0.078555  \n",
       "2                              0.0       0.082023  \n",
       "3                              0.0       0.047006  \n",
       "4                              0.0       0.092777  \n",
       "..                             ...            ...  \n",
       "858                            0.0       0.015587  \n",
       "859                            0.0       0.045282  \n",
       "860                            0.0       0.035870  \n",
       "861                            0.0       0.033834  \n",
       "862                            0.0       0.000574  \n",
       "\n",
       "[863 rows x 20 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9d2061bc-6f75-4a28-8156-baf3060b9db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df.values[cytobands_df.index[70]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc6cd1d0-454c-4354-9c18-bd9354568385",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sumOf_columns(df, colums=[\"CDSIntron_density\", \"UTR5Intron_density\", \"UTR3Intron_density\"], newColName = \"Introns_density\"):\n",
    "    \"\"\"\n",
    "    colums is a list of string\n",
    "    \"\"\"\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    new_densites = []\n",
    "    for row in df.index:\n",
    "        density =0\n",
    "        for column in colums:\n",
    "            density += df[column][row]\n",
    "        new_densites.append(round(density, 3))\n",
    "        \n",
    "    df.insert(len(df.columns), newColName, new_densites, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931407ff-fdfa-4df4-8ea6-cb2ab76afe52",
   "metadata": {},
   "outputs": [],
   "source": [
    "sumOf_columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8bf14c9-0b59-42bc-8b73-eec9e7cb7e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "586f8eb0-e8fd-4241-8e14-6da5756f8d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2212fb36-b602-4e07-be2a-5892afb6ada6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=CDS:ENSP00000493376;Parent=transcript:ENST00000641515;protein_id=ENSP00000493376'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDSs[\"Identifiers\"][99]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "129fc67e-4dbf-4f56-ab08-50f83017e28b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b8764313-5f77-4adc-98d1-50d81e0ed8b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \n",
       "0             0.012               0.025             0.019          0.171  \n",
       "1             0.004               0.013             0.006          0.079  \n",
       "2             0.006               0.012             0.010          0.082  \n",
       "3             0.005               0.004             0.004          0.047  \n",
       "4             0.005               0.014             0.010          0.093  \n",
       "..              ...                 ...               ...            ...  \n",
       "858           0.001               0.001             0.003          0.016  \n",
       "859           0.000               0.004             0.002          0.045  \n",
       "860           0.001               0.005             0.002          0.036  \n",
       "861           0.001               0.003             0.001          0.034  \n",
       "862           0.000               0.000             0.000          0.001  \n",
       "\n",
       "[863 rows x 18 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Anterior resultado com calculo de densidades com falhas, sobreposições devido às somas entre elementos diferentes na mesma matrix"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
