{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a0d93c-1359-4107-807f-f230b3863708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89dac285-5db9-4261-b4e4-38e0f84db137",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0       1                  2           3            4        5  6  7  \\\n",
       "0        1  GRCh38         chromosome         1.0  248956422.0        .  .  .   \n",
       "1        1       .  biological_region     10469.0      11240.0  1.3e+03  .  .   \n",
       "2        1       .  biological_region     10650.0      10657.0    0.999  +  .   \n",
       "3        1       .  biological_region     10655.0      10657.0    0.999  -  .   \n",
       "4        1       .  biological_region     10678.0      10687.0    0.999  +  .   \n",
       "...     ..     ...                ...         ...          ...      ... .. ..   \n",
       "3516515  Y       .  biological_region  57215804.0   57215813.0    0.999  -  .   \n",
       "3516516  Y       .  biological_region  57215832.0   57215840.0    0.999  +  .   \n",
       "3516517  Y       .  biological_region  57215833.0   57215842.0    0.999  -  .   \n",
       "3516518  Y       .  biological_region  57215862.0   57215867.0    0.999  +  .   \n",
       "3516519  Y       .  biological_region  57215862.0   57215870.0    0.999  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1                 external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2                                       logic_name=eponine  \n",
       "3                                       logic_name=eponine  \n",
       "4                                       logic_name=eponine  \n",
       "...                                                    ...  \n",
       "3516515                                 logic_name=eponine  \n",
       "3516516                                 logic_name=eponine  \n",
       "3516517                                 logic_name=eponine  \n",
       "3516518                                 logic_name=eponine  \n",
       "3516519                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_read= \"/home/androx/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens.GRCh38.112.chr.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f9135a3-08a2-46ed-9c66-2622790d4290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3516515   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3516516   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3516517   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3516518   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3516519   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                        Identifiers  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3516515      -       .                                 logic_name=eponine  \n",
       "3516516      +       .                                 logic_name=eponine  \n",
       "3516517      -       .                                 logic_name=eponine  \n",
       "3516518      +       .                                 logic_name=eponine  \n",
       "3516519      -       .                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Signal', 7: 'Unknown', 8: 'Identifiers'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64551187-3a83-4376-a7db-f0cc87b190b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ef8d7245-d6be-4015-bce8-07cb7093450f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328201</th>\n",
       "      <td>10</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133797422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467767</th>\n",
       "      <td>11</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>135086622.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673000</th>\n",
       "      <td>12</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133275309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862634</th>\n",
       "      <td>13</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>114364328.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924501</th>\n",
       "      <td>14</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>107043718.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048326</th>\n",
       "      <td>15</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101991189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171301</th>\n",
       "      <td>16</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90338345.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329042</th>\n",
       "      <td>17</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83257441.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536644</th>\n",
       "      <td>18</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80373285.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598313</th>\n",
       "      <td>19</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58617616.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791801</th>\n",
       "      <td>2</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242193529.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063156</th>\n",
       "      <td>20</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64444167.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145930</th>\n",
       "      <td>21</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46709983.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186784</th>\n",
       "      <td>22</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50818468.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2261287</th>\n",
       "      <td>3</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>198295559.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484781</th>\n",
       "      <td>4</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>190214555.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628956</th>\n",
       "      <td>5</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>181538259.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786252</th>\n",
       "      <td>6</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170805979.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2952608</th>\n",
       "      <td>7</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>159345973.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3120542</th>\n",
       "      <td>8</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145138636.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248106</th>\n",
       "      <td>9</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>138394717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384196</th>\n",
       "      <td>MT</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384360</th>\n",
       "      <td>X</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>156040895.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3505822</th>\n",
       "      <td>Y</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>57227415.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index   Last_index Value Signal  \\\n",
       "0         1  GRCh38  chromosome          1.0  248956422.0     .      .   \n",
       "328201   10  GRCh38  chromosome          1.0  133797422.0     .      .   \n",
       "467767   11  GRCh38  chromosome          1.0  135086622.0     .      .   \n",
       "673000   12  GRCh38  chromosome          1.0  133275309.0     .      .   \n",
       "862634   13  GRCh38  chromosome          1.0  114364328.0     .      .   \n",
       "924501   14  GRCh38  chromosome          1.0  107043718.0     .      .   \n",
       "1048326  15  GRCh38  chromosome          1.0  101991189.0     .      .   \n",
       "1171301  16  GRCh38  chromosome          1.0   90338345.0     .      .   \n",
       "1329042  17  GRCh38  chromosome          1.0   83257441.0     .      .   \n",
       "1536644  18  GRCh38  chromosome          1.0   80373285.0     .      .   \n",
       "1598313  19  GRCh38  chromosome          1.0   58617616.0     .      .   \n",
       "1791801   2  GRCh38  chromosome          1.0  242193529.0     .      .   \n",
       "2063156  20  GRCh38  chromosome          1.0   64444167.0     .      .   \n",
       "2145930  21  GRCh38  chromosome          1.0   46709983.0     .      .   \n",
       "2186784  22  GRCh38  chromosome          1.0   50818468.0     .      .   \n",
       "2261287   3  GRCh38  chromosome          1.0  198295559.0     .      .   \n",
       "2484781   4  GRCh38  chromosome          1.0  190214555.0     .      .   \n",
       "2628956   5  GRCh38  chromosome          1.0  181538259.0     .      .   \n",
       "2786252   6  GRCh38  chromosome          1.0  170805979.0     .      .   \n",
       "2952608   7  GRCh38  chromosome          1.0  159345973.0     .      .   \n",
       "3120542   8  GRCh38  chromosome          1.0  145138636.0     .      .   \n",
       "3248106   9  GRCh38  chromosome          1.0  138394717.0     .      .   \n",
       "3384196  MT  GRCh38  chromosome          1.0      16569.0     .      .   \n",
       "3384360   X  GRCh38  chromosome          1.0  156040895.0     .      .   \n",
       "3505822   Y  GRCh38  chromosome          1.0   57227415.0     .      .   \n",
       "\n",
       "        Unknown                                        Identifiers  \n",
       "0             .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "328201        .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  \n",
       "467767        .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  \n",
       "673000        .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  \n",
       "862634        .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  \n",
       "924501        .  ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...  \n",
       "1048326       .  ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...  \n",
       "1171301       .  ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...  \n",
       "1329042       .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...  \n",
       "1536644       .  ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...  \n",
       "1598313       .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...  \n",
       "1791801       .  ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...  \n",
       "2063156       .  ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...  \n",
       "2145930       .  ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...  \n",
       "2186784       .  ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...  \n",
       "2261287       .  ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...  \n",
       "2484781       .  ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...  \n",
       "2628956       .  ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...  \n",
       "2786252       .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...  \n",
       "2952608       .  ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...  \n",
       "3120542       .  ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...  \n",
       "3248106       .  ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...  \n",
       "3384196       .   ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1  \n",
       "3384360       .  ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...  \n",
       "3505822       .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "chromosmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0da5b2aa-931d-4e23-be59-d5b3e2b8e82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944575.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3515710</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25062548.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3515921</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516228</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57067865.0</td>\n",
       "      <td>57130289.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516327</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57184216.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516392</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57207346.0</td>\n",
       "      <td>57212230.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21547 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "93        1  ensembl_havana  gene      65419.0     71585.0     .      +   \n",
       "395       1  ensembl_havana  gene     450740.0    451678.0     .      -   \n",
       "541       1  ensembl_havana  gene     685716.0    686654.0     .      -   \n",
       "1161      1  ensembl_havana  gene     923923.0    944575.0     .      +   \n",
       "1524      1  ensembl_havana  gene     944203.0    959309.0     .      -   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3515710   Y  ensembl_havana  gene   25030901.0  25062548.0     .      -   \n",
       "3515921   Y  ensembl_havana  gene   25622117.0  25624902.0     .      +   \n",
       "3516228   Y  ensembl_havana  gene   57067865.0  57130289.0     .      +   \n",
       "3516327   Y  ensembl_havana  gene   57184216.0  57197337.0     .      +   \n",
       "3516392   Y  ensembl_havana  gene   57207346.0  57212230.0     .      +   \n",
       "\n",
       "        Unknown                                        Identifiers  \n",
       "93            .  ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...  \n",
       "395           .  ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...  \n",
       "541           .  ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...  \n",
       "1161          .  ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...  \n",
       "1524          .  ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...  \n",
       "...         ...                                                ...  \n",
       "3515710       .  ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...  \n",
       "3515921       .  ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...  \n",
       "3516228       .  ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...  \n",
       "3516327       .  ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...  \n",
       "3516392       .  ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...  \n",
       "\n",
       "[21547 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2229627f-149a-46f7-9a75-93c56ca4b2b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "93         ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...\n",
       "395        ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...\n",
       "541        ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...\n",
       "1161       ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...\n",
       "1524       ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...\n",
       "                                 ...                        \n",
       "3515710    ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...\n",
       "3515921    ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...\n",
       "3516228    ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...\n",
       "3516327    ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...\n",
       "3516392    ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...\n",
       "Name: Identifiers, Length: 21547, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[\"Identifiers\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "47ddd434-fd68-4235-8d5c-7c69130968c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565.0</td>\n",
       "      <td>65573.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037.0</td>\n",
       "      <td>70008.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000409316;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000329982;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1165</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432.0</td>\n",
       "      <td>924948.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000478421;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516411</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209532.0</td>\n",
       "      <td>57209733.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516413</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209822.0</td>\n",
       "      <td>57209980.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516415</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57210640.0</td>\n",
       "      <td>57210792.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516417</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211552.0</td>\n",
       "      <td>57211620.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516419</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211761.0</td>\n",
       "      <td>57211894.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900145 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome Type  First_index  Last_index Value Signal  \\\n",
       "99        1          havana  CDS      65565.0     65573.0     .      +   \n",
       "100       1          havana  CDS      69037.0     70008.0     .      +   \n",
       "398       1  ensembl_havana  CDS     450740.0    451678.0     .      -   \n",
       "544       1  ensembl_havana  CDS     685716.0    686654.0     .      -   \n",
       "1165      1          havana  CDS     924432.0    924948.0     .      +   \n",
       "...      ..             ...  ...          ...         ...   ...    ...   \n",
       "3516411   Y          havana  CDS   57209532.0  57209733.0     .      +   \n",
       "3516413   Y          havana  CDS   57209822.0  57209980.0     .      +   \n",
       "3516415   Y          havana  CDS   57210640.0  57210792.0     .      +   \n",
       "3516417   Y          havana  CDS   57211552.0  57211620.0     .      +   \n",
       "3516419   Y          havana  CDS   57211761.0  57211894.0     .      +   \n",
       "\n",
       "        Unknown                                        Identifiers  \n",
       "99            0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "100           0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "398           0  ID=CDS:ENSP00000409316;Parent=transcript:ENST0...  \n",
       "544           0  ID=CDS:ENSP00000329982;Parent=transcript:ENST0...  \n",
       "1165          0  ID=CDS:ENSP00000478421;Parent=transcript:ENST0...  \n",
       "...         ...                                                ...  \n",
       "3516411       0  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516413       2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516415       2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516417       2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516419       2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "\n",
       "[900145 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "CDSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8af64bb5-1b2d-4493-80a0-939582f088b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516493</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212184.0</td>\n",
       "      <td>57213125.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516494</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213204.0</td>\n",
       "      <td>57213357.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516495</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213526.0</td>\n",
       "      <td>57213602.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516496</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213880.0</td>\n",
       "      <td>57213964.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516497</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214350.0</td>\n",
       "      <td>57214397.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1668627 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome  Type  First_index  Last_index Value Signal Unknown  \\\n",
       "18        1  havana  exon      11869.0     12227.0     .      +       .   \n",
       "19        1  havana  exon      12613.0     12721.0     .      +       .   \n",
       "20        1  havana  exon      13221.0     14409.0     .      +       .   \n",
       "24        1  havana  exon      12010.0     12057.0     .      +       .   \n",
       "25        1  havana  exon      12179.0     12227.0     .      +       .   \n",
       "...      ..     ...   ...          ...         ...   ...    ...     ...   \n",
       "3516493   Y  havana  exon   57212184.0  57213125.0     .      -       .   \n",
       "3516494   Y  havana  exon   57213204.0  57213357.0     .      -       .   \n",
       "3516495   Y  havana  exon   57213526.0  57213602.0     .      -       .   \n",
       "3516496   Y  havana  exon   57213880.0  57213964.0     .      -       .   \n",
       "3516497   Y  havana  exon   57214350.0  57214397.0     .      -       .   \n",
       "\n",
       "                                               Identifiers  \n",
       "18       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "19       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "20       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "24       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "25       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "...                                                    ...  \n",
       "3516493  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516494  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516495  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516496  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516497  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "\n",
       "[1668627 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3b61b919-6b9b-47a9-a7de-dfa3e64dbd80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70009.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1223</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000618323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000342066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154.0</td>\n",
       "      <td>944575.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000341065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516273</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57128403.0</td>\n",
       "      <td>57128464.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516298</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57128660.0</td>\n",
       "      <td>57130289.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516315</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57128472.0</td>\n",
       "      <td>57128509.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516348</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57196930.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516369</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57196373.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>211648 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome             Type  First_index  Last_index Value  \\\n",
       "102       1          havana  three_prime_UTR      70009.0     71585.0     .   \n",
       "1192      1          havana  three_prime_UTR     944154.0    944574.0     .   \n",
       "1223      1          havana  three_prime_UTR     944154.0    944574.0     .   \n",
       "1266      1          havana  three_prime_UTR     944154.0    944574.0     .   \n",
       "1415      1          havana  three_prime_UTR     944154.0    944575.0     .   \n",
       "...      ..             ...              ...          ...         ...   ...   \n",
       "3516273   Y          havana  three_prime_UTR   57128403.0  57128464.0     .   \n",
       "3516298   Y  ensembl_havana  three_prime_UTR   57128660.0  57130289.0     .   \n",
       "3516315   Y  ensembl_havana  three_prime_UTR   57128472.0  57128509.0     .   \n",
       "3516348   Y  ensembl_havana  three_prime_UTR   57196930.0  57197337.0     .   \n",
       "3516369   Y  ensembl_havana  three_prime_UTR   57196373.0  57197337.0     .   \n",
       "\n",
       "        Signal Unknown                        Identifiers  \n",
       "102          +       .  Parent=transcript:ENST00000641515  \n",
       "1192         +       .  Parent=transcript:ENST00000616016  \n",
       "1223         +       .  Parent=transcript:ENST00000618323  \n",
       "1266         +       .  Parent=transcript:ENST00000342066  \n",
       "1415         +       .  Parent=transcript:ENST00000341065  \n",
       "...        ...     ...                                ...  \n",
       "3516273      +       .  Parent=transcript:ENST00000711262  \n",
       "3516298      +       .  Parent=transcript:ENST00000711264  \n",
       "3516315      +       .  Parent=transcript:ENST00000711265  \n",
       "3516348      +       .  Parent=transcript:ENST00000711286  \n",
       "3516369      +       .  Parent=transcript:ENST00000711287  \n",
       "\n",
       "[211648 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR3 = tanscripts_df.loc[tanscripts_df['Type'] == \"three_prime_UTR\"]\n",
    "UTR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ed7c774-ed75-4a23-9548-47d0a9913bcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>65433.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>65520.0</td>\n",
       "      <td>65564.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>924431.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>924431.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000618323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1226</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>925150.0</td>\n",
       "      <td>925189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000437963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516285</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57075978.0</td>\n",
       "      <td>57075986.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516301</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57067898.0</td>\n",
       "      <td>57067968.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516302</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57075978.0</td>\n",
       "      <td>57075986.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516329</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57184216.0</td>\n",
       "      <td>57184279.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516350</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57184226.0</td>\n",
       "      <td>57184263.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175417 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome            Type  First_index  Last_index Value  \\\n",
       "96        1          havana  five_prime_UTR      65419.0     65433.0     .   \n",
       "97        1          havana  five_prime_UTR      65520.0     65564.0     .   \n",
       "1163      1          havana  five_prime_UTR     923923.0    924431.0     .   \n",
       "1194      1          havana  five_prime_UTR     923923.0    924431.0     .   \n",
       "1226      1          havana  five_prime_UTR     925150.0    925189.0     .   \n",
       "...      ..             ...             ...          ...         ...   ...   \n",
       "3516285   Y  ensembl_havana  five_prime_UTR   57075978.0  57075986.0     .   \n",
       "3516301   Y  ensembl_havana  five_prime_UTR   57067898.0  57067968.0     .   \n",
       "3516302   Y  ensembl_havana  five_prime_UTR   57075978.0  57075986.0     .   \n",
       "3516329   Y  ensembl_havana  five_prime_UTR   57184216.0  57184279.0     .   \n",
       "3516350   Y  ensembl_havana  five_prime_UTR   57184226.0  57184263.0     .   \n",
       "\n",
       "        Signal Unknown                        Identifiers  \n",
       "96           +       .  Parent=transcript:ENST00000641515  \n",
       "97           +       .  Parent=transcript:ENST00000641515  \n",
       "1163         +       .  Parent=transcript:ENST00000616016  \n",
       "1194         +       .  Parent=transcript:ENST00000618323  \n",
       "1226         +       .  Parent=transcript:ENST00000437963  \n",
       "...        ...     ...                                ...  \n",
       "3516285      +       .  Parent=transcript:ENST00000711264  \n",
       "3516301      +       .  Parent=transcript:ENST00000711265  \n",
       "3516302      +       .  Parent=transcript:ENST00000711265  \n",
       "3516329      +       .  Parent=transcript:ENST00000711286  \n",
       "3516350      +       .  Parent=transcript:ENST00000711287  \n",
       "\n",
       "[175417 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR5 = tanscripts_df.loc[tanscripts_df['Type'] == \"five_prime_UTR\"]\n",
    "UTR5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3b61a75d-2316-4377-9a54-a7ad6b86b487",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "867e6833-f85a-473a-add0-e59b583c5f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_matrix(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    The matrix representing all intances of a sequence type available from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR \n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            chr_array[int(chr_sequence[\"First_index\"][row]) : int(chr_sequence[\"Last_index\"][row])] += 1\n",
    "                \n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "741ead14-c552-448e-96f7-2fd45c75f9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def introns_exons_matrix(chromosomes_df, sequence1_df, exons_df):\n",
    "    \"\"\"\n",
    "    sequence1 = CDS_df, 5 prime UTR or 3 prime UTR\n",
    "    if sequence1_df: +2\n",
    "    if exon: +1\n",
    "    introns = 2\n",
    "    exons = 3\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence1_df.loc[sequence1_df['Chr'] == chromosome]         #can be CDS, 5 prime UTR or 3 prime UTR\n",
    "        chr_exons = exons_df.loc[exons_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        #CDS region\n",
    "        for row in chr_sequence.index:\n",
    "            chr_array[int(chr_sequence[\"First_index\"][row]): int(chr_sequence[\"Last_index\"][row])] += 2\n",
    "        #exons region\n",
    "        for row in chr_exons.index:\n",
    "            chr_array[int(chr_exons[\"First_index\"][row]): int(chr_exons[\"Last_index\"][row])] += 1\n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0053bc01-6bdd-46e1-9592-41e0f5a2599b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity(cytobands_df, element_matrix, col_name=\"Element_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    densities = []\n",
    "    chromosomes = cytobands_df[\"Chromossome\"].unique()\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            cytoband_array = element_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            dr_length = np.count_nonzero(cytoband_array)\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            density = dr_length/cytoband_length\n",
    "            densities.append(round(density, 3))\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9e6e7d8a-9c5d-4c16-95ba-720845fc909c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity(cytobands_df, introns_matrix, col_name1=\"intron_density\", col_name2=\"exon_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    chromosomes = cytobands_df[\"Chromossome\"].unique()\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            cytoband_array = introns_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            intron_length = np.count_nonzero(cytoband_array == 2)         #para introns\n",
    "            exon_length = np.count_nonzero(cytoband_array == 3)         #para exons\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row] \n",
    "            intron_density = intron_length/cytoband_length\n",
    "            exon_density = exon_length/cytoband_length\n",
    "            intron_densities.append(round(intron_density, 3))\n",
    "            exon_densities.append(round(exon_density, 3))\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name1, intron_densities, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name2, exon_densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "84e048d0-2f72-453e-bdb7-d9007893a020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    df_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        df_cols += col + \"\\t\"\n",
    "    df_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(df_cols))\n",
    "    for row in df.values:\n",
    "        df_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            df_row += str(col) + \"\\t\"\n",
    "        df_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(df_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd457d08-3d9c-4ad7-b34a-e4cdd5aa389d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  \n",
       "0          0.088       0.326  \n",
       "1          0.084       0.282  \n",
       "2          0.056       0.275  \n",
       "3          0.049       0.405  \n",
       "4          0.056       0.404  \n",
       "..           ...         ...  \n",
       "858        0.037       0.015  \n",
       "859        0.062       0.013  \n",
       "860        0.082       0.006  \n",
       "861        0.026       0.002  \n",
       "862        0.006       0.001  \n",
       "\n",
       "[863 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoPath_read = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoPath_write = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr_dg.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoPath_read, sep=\"\\t\")#, header=None))\n",
    "cytobands_df2 = pd.read_csv(cytoPath_write, sep=\"\\t\")\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7479238e-b553-4e99-b933-9b39a635e849",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes_matrix = element_matrix(chromosmes, genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "16a3bcff-05c6-47e6-973c-88e24425e68c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df, genes_matrix, \"genes_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1e8cdd-042e-4071-9737-69169c01ee2d",
   "metadata": {},
   "source": [
    "# Code running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5377fb6f-0db3-40d0-9c18-1632ee187249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \n",
       "0             0.012               0.025             0.019          0.171  \n",
       "1             0.004               0.013             0.006          0.079  \n",
       "2             0.006               0.012             0.010          0.082  \n",
       "3             0.005               0.004             0.004          0.047  \n",
       "4             0.005               0.014             0.010          0.093  \n",
       "..              ...                 ...               ...            ...  \n",
       "858           0.001               0.001             0.003          0.016  \n",
       "859           0.000               0.004             0.002          0.045  \n",
       "860           0.001               0.005             0.002          0.036  \n",
       "861           0.001               0.003             0.001          0.034  \n",
       "862           0.000               0.000             0.000          0.001  \n",
       "\n",
       "[863 rows x 18 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9495e9-9702-4571-a78f-9d4835b82301",
   "metadata": {},
   "source": [
    "## CDS density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3c8f619e-9ac7-4170-9ec3-b2b1a2303fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CDSs_matrix = element_matrix(chromosmes, CDSs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d65c4c8c-7830-4ae4-928f-2104138ddc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df, CDSs_matrix, \"CDSs_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b4e30cb7-2f45-4a74-8175-fac4802502fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d690d7f-d2aa-453e-b132-1ac6e85dcf68",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within CDSs (Coding Sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f257b5cd-3b3b-4d28-8f80-2000a6263843",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CDS_introns_exons_matrix = introns_exons_matrix(chromosmes, CDSs, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6303e000-b492-488a-9f78-3b43abd095a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df, CDS_introns_exons_matrix, \"CDSIntron_density\", \"CDSExon_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "55598914-f945-4356-8645-7eda2944190d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b149728d-101f-4b84-9fa8-a77b5686e219",
   "metadata": {},
   "source": [
    "## Prime 5 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d4d6dc2d-563f-4088-88c3-d623f062ff1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_matrix = element_matrix(chromosmes, UTR5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9737e5af-4410-42d5-a49f-2a19b76d6e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df, UTR5_matrix, \"5P_UTR_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2778db1a-3cd0-46d2-961c-4a3b807334a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bd2ce3-76ac-481a-a4af-20733256d243",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within 5 Prime UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "760c6bd2-e12c-43f9-b93b-8ab3980437b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_introns_exons_matrix = introns_exons_matrix(chromosmes, UTR5, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2a1e58ff-3d9e-4b90-b0cc-217029d973af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df2, UTR5_introns_exons_matrix, \"UTR5Intron_density\", \"UTR5Exon_density\")\n",
    "#cytobands_df\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e3aad2b6-cdde-4385-91e8-db90f76dd848",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51f8d5e-d87f-47bc-89ef-60478ed00610",
   "metadata": {},
   "source": [
    "## Prime 3 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2b70b50a-9eae-4fa6-932b-9ce572f7fd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR3_matrix = element_matrix(chromosmes, UTR3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a748357c-732c-4307-81fe-c5e28054425a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df2, UTR3_matrix, \"3P_UTR_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "06edbc35-0c00-46c2-90e2-b9b14eeacd4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d99a6f99-a946-4123-8f99-8984c9b011a0",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within 3 Prime UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a6f4ede0-3475-415d-995e-5096c30268f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR3_introns_exons_matrix = introns_exons_matrix(chromosmes, UTR3, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "33f5773d-6a69-4e55-a936-f88ee3999716",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df2, UTR3_introns_exons_matrix, \"UTR3Intron_density\", \"UTR3Exon_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "383b40f1-7696-41e8-bb07-3dceca86bd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc99fd3-b968-4579-9ce9-c0ceec7f6925",
   "metadata": {},
   "source": [
    "## Exons density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b8e17263-abd9-4634-b815-4ef21afd7bf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "#exons_matrix = element_matrix(chromosmes, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b8764313-5f77-4adc-98d1-50d81e0ed8b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \n",
       "0             0.012               0.025             0.019          0.171  \n",
       "1             0.004               0.013             0.006          0.079  \n",
       "2             0.006               0.012             0.010          0.082  \n",
       "3             0.005               0.004             0.004          0.047  \n",
       "4             0.005               0.014             0.010          0.093  \n",
       "..              ...                 ...               ...            ...  \n",
       "858           0.001               0.001             0.003          0.016  \n",
       "859           0.000               0.004             0.002          0.045  \n",
       "860           0.001               0.005             0.002          0.036  \n",
       "861           0.001               0.003             0.001          0.034  \n",
       "862           0.000               0.000             0.000          0.001  \n",
       "\n",
       "[863 rows x 18 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calc_elementDensity(cytobands_df2, exons_matrix, \"Exons_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6bbc1ed2-3117-4308-b855-aae88c549928",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9d2061bc-6f75-4a28-8156-baf3060b9db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df.values[cytobands_df.index[70]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc6cd1d0-454c-4354-9c18-bd9354568385",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sumOf_columns(df, colums=[\"CDSIntron_density\", \"UTR5Intron_density\", \"UTR3Intron_density\"], newColName = \"Introns_density\"):\n",
    "    \"\"\"\n",
    "    colums is a list of string\n",
    "    \"\"\"\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    new_densites = []\n",
    "    for row in df.index:\n",
    "        density =0\n",
    "        for column in colums:\n",
    "            density += df[column][row]\n",
    "        new_densites.append(round(density, 3))\n",
    "        \n",
    "    df.insert(len(df.columns), newColName, new_densites, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c3c720a7-b910-442a-9125-3bfb4205f7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "      <th>Introns_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "      <td>0.069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \\\n",
       "0             0.012               0.025             0.019          0.171   \n",
       "1             0.004               0.013             0.006          0.079   \n",
       "2             0.006               0.012             0.010          0.082   \n",
       "3             0.005               0.004             0.004          0.047   \n",
       "4             0.005               0.014             0.010          0.093   \n",
       "..              ...                 ...               ...            ...   \n",
       "858           0.001               0.001             0.003          0.016   \n",
       "859           0.000               0.004             0.002          0.045   \n",
       "860           0.001               0.005             0.002          0.036   \n",
       "861           0.001               0.003             0.001          0.034   \n",
       "862           0.000               0.000             0.000          0.001   \n",
       "\n",
       "     Introns_density  \n",
       "0              0.069  \n",
       "1              0.034  \n",
       "2              0.031  \n",
       "3              0.011  \n",
       "4              0.034  \n",
       "..               ...  \n",
       "858            0.003  \n",
       "859            0.012  \n",
       "860            0.013  \n",
       "861            0.008  \n",
       "862            0.000  \n",
       "\n",
       "[863 rows x 19 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sumOf_columns(cytobands_df2)\n",
    "cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8bf14c9-0b59-42bc-8b73-eec9e7cb7e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586f8eb0-e8fd-4241-8e14-6da5756f8d72",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
