{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a0d93c-1359-4107-807f-f230b3863708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89dac285-5db9-4261-b4e4-38e0f84db137",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_read= \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens.GRCh38.112.chr.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "#tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f9135a3-08a2-46ed-9c66-2622790d4290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3516515   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3516516   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3516517   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3516518   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3516519   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                        Identifiers  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3516515      -       .                                 logic_name=eponine  \n",
       "3516516      +       .                                 logic_name=eponine  \n",
       "3516517      -       .                                 logic_name=eponine  \n",
       "3516518      +       .                                 logic_name=eponine  \n",
       "3516519      -       .                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Signal', 7: 'Unknown', 8: 'Identifiers'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64551187-3a83-4376-a7db-f0cc87b190b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c35e5d62-69a8-44a0-be89-734bae0489bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=chromosome:1;Alias=CM000663.2,chr1,NC_000001.11'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Identifiers\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8d7245-d6be-4015-bce8-07cb7093450f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "#genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "#CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "#UTR3 = tanscripts_df.loc[tanscripts_df['Type'] == \"three_prime_UTR\"]\n",
    "#UTR5 = tanscripts_df.loc[tanscripts_df['Type'] == \"five_prime_UTR\"]\n",
    "#exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb5dde82-4b0a-4b05-a2fa-e537ba1940bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR3[\"Identifiers\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "49bbd252-0fed-472e-a11c-f5eba8d3f97c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_overlap = UTR5.loc[UTR5[\"Identifiers\"] == UTR5[\"Identifiers\"][3516301]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62ea00e7-0a4d-4a13-936f-02650f2bdb54",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_overlap.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "86fb3cf3-3746-4a7b-ae73-8dafdbf4be72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_NoOverlaps(element_type = \"CDS\"):\n",
    "    element_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    elements_dict = {}\n",
    "    #elements_dict.update({\"header\": element_df.columns.tolist()})\n",
    "    count = 0\n",
    "    for row in element_df.index:\n",
    "        identifiers = element_df[\"Identifiers\"][row]\n",
    "        transcript_id = identifiers.split(\";\")[0]\n",
    "        transcript_length = element_df[\"Last_index\"][row] - element_df[\"First_index\"][row]\n",
    "        if transcript_id not in elements_dict.keys():\n",
    "        #if elements_dict.get(transcript_id) == None:\n",
    "            #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "            elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        else:\n",
    "            #print(\"resolving overlap...\")\n",
    "            if transcript_length > elements_dict[transcript_id][1]:\n",
    "                #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "                elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        count+=1\n",
    "\n",
    "    #transcripts_matrix = []\n",
    "    #for value in elements_dict.values():\n",
    "    #    transcripts_matrix.append(value[0])\n",
    "    return elements_dict\n",
    "    #return pd.DataFrame(elements_dict.values())\n",
    "    #return pd.DataFrame(transcripts_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cd7ce5a1-dd7c-445d-8388-e423544af64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_dict_toDF(dictionary):\n",
    "    transcripts_matrix = []\n",
    "    concat_list = []\n",
    "    for value in dictionary.values():\n",
    "        concat_list.append(value[0])\n",
    "    return pd.concat(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a1b2b71-b9c0-4267-9e61-2a142000233f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fivePrime_UTR_dict = transcripts_NoOverlaps(\"five_prime_UTR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7a064b-470e-412d-b993-49f90c35aa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for value in fivePrime_UTR_dict.values():\n",
    "#    print(type(value[0]))\n",
    "#    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0006eaa1-0cac-4ed6-bb12-05c631af2605",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df = transcripts_dict_toDF(fivePrime_UTR_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0554414-1b4b-4fd1-8fd6-9d1269d206f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171e9519-2e5a-4085-82e0-8b2b12b53fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_noOverlaps_df[\"Identifiers\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "867e6833-f85a-473a-add0-e59b583c5f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_matrix(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    The matrix representing all intances of a sequence type available from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            chr_array[int(chr_sequence[\"First_index\"][row]) : int(chr_sequence[\"Last_index\"][row])] += 1\n",
    "                \n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0053bc01-6bdd-46e1-9592-41e0f5a2599b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity(cytobands_df, region_matrix, col_name=\"Element_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    densities = []\n",
    "    chromosomes = cytobands_df[\"Chromossome\"].unique()\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            dr_length = np.count_nonzero(cytoband_array)\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            density = dr_length/cytoband_length\n",
    "            densities.append(round(density, 10))\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9e6e7d8a-9c5d-4c16-95ba-720845fc909c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity(cytobands_df, region_matrix, exons_matrix, col_name1=\"exon_density\", col_name2=\"intron_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    chromosomes = cytobands_df[\"Chromossome\"].unique()\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):\n",
    "        print(\"Calculating densities in chromosome: \", cytobands_df[\"Chromossome\"].unique()[chrom], \" ...\")\n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            ############################ Resolver problema, o conjunto dos exons contem todo o conjunto dos CDSs ###############################\n",
    "            region_cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            exon_cytoband_array = exons_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            region_indexes = np.where(region_cytoband_array != 0)[0].tolist()\n",
    "            exon_indexes = np.where(exon_cytoband_array != 0)[0].tolist()\n",
    "            print(\"exons in CDS region: \", np.intersect1d(region_indexes, exon_indexes))\n",
    "            print(\"introns in CDS region: \", np.setdiff1d(region_indexes, exon_indexes))\n",
    "            exon_length = len(np.intersect1d(region_indexes, exon_indexes))    #indices que estam na região e são exões\n",
    "            intron_length = len(np.setdiff1d(region_indexes, exon_indexes))      #indices que estam na região mas não são exões\n",
    "            #for region_element in region_indexes:\n",
    "            #    if region_element in exon_indexes:\n",
    "            #        exon_length+=1\n",
    "            #    elif region_element not in exon_indexes:\n",
    "            #        intron_length+=1\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            exon_density = exon_length/cytoband_length\n",
    "            intron_density = intron_length/cytoband_length\n",
    "            exon_densities.append(round(exon_density, 10))\n",
    "            intron_densities.append(round(intron_density, 10))   #max 6\n",
    "            \n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name1, exon_densities, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name2, intron_densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "93fd572c-e059-4592-8147-cbeb4f16de06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2 3 7 8]\n",
      "[1 6]\n"
     ]
    }
   ],
   "source": [
    "array1 = [1, 2, 3, 6, 7, 8]\n",
    "array2 = [2, 3, 4, 7, 8, 9]\n",
    "print(np.intersect1d(array1, array2))\n",
    "print(np.setdiff1d(array1, array2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84e048d0-2f72-453e-bdb7-d9007893a020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    df_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        df_cols += col + \"\\t\"\n",
    "    df_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(df_cols))\n",
    "    for row in df.values:\n",
    "        df_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            df_row += str(col) + \"\\t\"\n",
    "        df_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(df_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd457d08-3d9c-4ad7-b34a-e4cdd5aa389d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  \n",
       "0          0.088       0.326  \n",
       "1          0.084       0.282  \n",
       "2          0.056       0.275  \n",
       "3          0.049       0.405  \n",
       "4          0.056       0.404  \n",
       "..           ...         ...  \n",
       "858        0.037       0.015  \n",
       "859        0.062       0.013  \n",
       "860        0.082       0.006  \n",
       "861        0.026       0.002  \n",
       "862        0.006       0.001  \n",
       "\n",
       "[863 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoPath_read = \"/home/andre/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoPath_write = \"/home/andre/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr_dg.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoPath_read, sep=\"\\t\")#, header=None))\n",
    "cytobands_df2 = pd.read_csv(cytoPath_write, sep=\"\\t\")\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e729f140-d836-4918-8b62-57d7b5b00e2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "UTR5 = tanscripts_df.loc[tanscripts_df['Type'] == \"five_prime_UTR\"]\n",
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ed8a71e4-c2b7-4d34-982c-7533796159d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "UTR5_matrix = element_matrix(chromosmes, UTR5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "16a3bcff-05c6-47e6-973c-88e24425e68c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>5P UTR_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.012022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.003561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.006496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.004600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.005499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.001027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.000346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.000675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  5P UTR_density  \n",
       "0          0.088       0.326        0.012022  \n",
       "1          0.084       0.282        0.003561  \n",
       "2          0.056       0.275        0.006496  \n",
       "3          0.049       0.405        0.004600  \n",
       "4          0.056       0.404        0.005499  \n",
       "..           ...         ...             ...  \n",
       "858        0.037       0.015        0.001027  \n",
       "859        0.062       0.013        0.000346  \n",
       "860        0.082       0.006        0.000675  \n",
       "861        0.026       0.002        0.000838  \n",
       "862        0.006       0.001        0.000006  \n",
       "\n",
       "[863 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calc_elementDensity(cytobands_df, UTR5_matrix, \"5P UTR_density\")\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7c95654a-c73f-4aad-9d24-04fcc01e7691",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "exon_matrix = element_matrix(chromosmes, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "57cd8058-6c32-4afa-9f82-33b97d57c7aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating densities in chromosome:  chr1  ...\n",
      "exons in CDS region:  [  65419   65420   65421 ... 2228763 2228764 2228765]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  91534   91535   91536 ... 2355422 2355423 2355424]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 686290  686291  686292 ... 1485547 1485548 1485549]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 569759  569760  569761 ... 1988475 1988476 1988477]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  29028   29029   29030 ... 3397628 3397629 3397630]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   6853    6854    6855 ... 3372933 3372934 3372935]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  47061   47062   47063 ... 4184827 4184828 4184829]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  19499   19500   19501 ... 3459498 3459499 3459500]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  91742   91743   91744 ... 3958880 3958881 3958882]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   3397    3398    3399 ... 1636641 1636642 1636643]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 823552  823553  823554 ... 2392166 2392167 2392168]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  36034   36035   36036 ... 1897471 1897472 1897473]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 455047  455048  455049 ... 5276787 5276788 5276789]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  39626   39627   39628 ... 4053172 4053173 4053174]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   5824    5825    5826 ... 2599887 2599888 2599889]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   3257    3258    3259 ... 3845112 3845113 3845114]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 221537  221538  221539 ... 5015375 5015376 5015377]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 979017  979018  979019 ... 1824057 1824058 1824059]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  39295   39296   39297 ... 1573767 1573768 1573769]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  65259   65260   65261 ... 7697218 7697219 7697220]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  []\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  267922   267923   267924 ... 15098561 15098562 15098563]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   8230    8231    8232 ... 2931813 2932013 2932014]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 784222  784223  784224 ... 3504853 3504854 3504855]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [    851     852     853 ... 2775065 2775066 2775067]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 118389  118390  118391 ... 4964590 4964591 4964592]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 346113  346114  346115 ... 1938981 1938982 1938983]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  30819   30820   30821 ... 1950466 1950467 1950468]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 356674  356675  356676 ... 4447711 4447712 4447713]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [    460     461     462 ... 4138767 4138768 4138769]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 141970  141971  141972 ... 1673263 1673264 1673265]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   3618    3619    3620 ... 2976517 2976518 2976519]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 15035  15036  15037 ... 965325 965326 965327]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  []\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  []\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  []\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [ 705967  705968  705969 ... 4052708 4052709 4052710]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  41501   41502   41503 ... 3079735 3079736 3079737]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  29106   29107   29108 ... 4478938 4478939 4478940]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  27876   27877   27878 ... 1499101 1499102 1499103]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [   1409    1410    1411 ... 2476763 2476764 2476765]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  22239   22240   22241 ... 1315834 1315835 1315836]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  23193   23194   23195 ... 4945352 4945353 4945354]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  44000   44001   44002 ... 1666476 1666477 1666478]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  20886   20887   20888 ... 3464215 3464216 3464217]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  35526   35527   35528 ... 1759198 1759199 1759200]\n",
      "introns in CDS region:  []\n",
      "exons in CDS region:  [  50897   50898   50899 ... 2743592 2743593 2743594]\n",
      "introns in CDS region:  []\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mcalc_ExonIntronDensity\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcytobands_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mUTR5_matrix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexon_matrix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcol_name1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mUTR5_exon_density\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcol_name2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mUTR5_intron_density\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m cytobands_df\n",
      "Cell \u001b[0;32mIn[19], line 15\u001b[0m, in \u001b[0;36mcalc_ExonIntronDensity\u001b[0;34m(cytobands_df, region_matrix, exons_matrix, col_name1, col_name2)\u001b[0m\n\u001b[1;32m     13\u001b[0m region_indexes \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mwhere(region_cytoband_array \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[1;32m     14\u001b[0m exon_indexes \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mwhere(exon_cytoband_array \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[0;32m---> 15\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexons in CDS region: \u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mintersect1d\u001b[49m\u001b[43m(\u001b[49m\u001b[43mregion_indexes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexon_indexes\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mintrons in CDS region: \u001b[39m\u001b[38;5;124m\"\u001b[39m, np\u001b[38;5;241m.\u001b[39msetdiff1d(region_indexes, exon_indexes))\n\u001b[1;32m     17\u001b[0m exon_length \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(np\u001b[38;5;241m.\u001b[39mintersect1d(region_indexes, exon_indexes))    \u001b[38;5;66;03m#indices que estam na região e são exões\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/numpy/lib/_arraysetops_impl.py:681\u001b[0m, in \u001b[0;36mintersect1d\u001b[0;34m(ar1, ar2, assume_unique, return_indices)\u001b[0m\n\u001b[1;32m    679\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    680\u001b[0m         ar1 \u001b[38;5;241m=\u001b[39m unique(ar1)\n\u001b[0;32m--> 681\u001b[0m         ar2 \u001b[38;5;241m=\u001b[39m \u001b[43munique\u001b[49m\u001b[43m(\u001b[49m\u001b[43mar2\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    682\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    683\u001b[0m     ar1 \u001b[38;5;241m=\u001b[39m ar1\u001b[38;5;241m.\u001b[39mravel()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/numpy/lib/_arraysetops_impl.py:291\u001b[0m, in \u001b[0;36munique\u001b[0;34m(ar, return_index, return_inverse, return_counts, axis, equal_nan)\u001b[0m\n\u001b[1;32m    289\u001b[0m ar \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masanyarray(ar)\n\u001b[1;32m    290\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m axis \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 291\u001b[0m     ret \u001b[38;5;241m=\u001b[39m \u001b[43m_unique1d\u001b[49m\u001b[43m(\u001b[49m\u001b[43mar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_inverse\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_counts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mequal_nan\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mequal_nan\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minverse_shape\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mar\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    293\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _unpack_tuple(ret)\n\u001b[1;32m    295\u001b[0m \u001b[38;5;66;03m# axis was specified and not None\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/numpy/lib/_arraysetops_impl.py:358\u001b[0m, in \u001b[0;36m_unique1d\u001b[0;34m(ar, return_index, return_inverse, return_counts, equal_nan, inverse_shape, axis)\u001b[0m\n\u001b[1;32m    356\u001b[0m     aux \u001b[38;5;241m=\u001b[39m ar[perm]\n\u001b[1;32m    357\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 358\u001b[0m     \u001b[43mar\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    359\u001b[0m     aux \u001b[38;5;241m=\u001b[39m ar\n\u001b[1;32m    360\u001b[0m mask \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty(aux\u001b[38;5;241m.\u001b[39mshape, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mbool)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "calc_ExonIntronDensity(cytobands_df, UTR5_matrix, exon_matrix, col_name1=\"UTR5_exon_density\", col_name2=\"UTR5_intron_density\")\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1e8cdd-042e-4071-9737-69169c01ee2d",
   "metadata": {},
   "source": [
    "# Code running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5377fb6f-0db3-40d0-9c18-1632ee187249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \n",
       "0             0.012               0.025             0.019          0.171  \n",
       "1             0.004               0.013             0.006          0.079  \n",
       "2             0.006               0.012             0.010          0.082  \n",
       "3             0.005               0.004             0.004          0.047  \n",
       "4             0.005               0.014             0.010          0.093  \n",
       "..              ...                 ...               ...            ...  \n",
       "858           0.001               0.001             0.003          0.016  \n",
       "859           0.000               0.004             0.002          0.045  \n",
       "860           0.001               0.005             0.002          0.036  \n",
       "861           0.001               0.003             0.001          0.034  \n",
       "862           0.000               0.000             0.000          0.001  \n",
       "\n",
       "[863 rows x 18 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6fb2715-301c-4988-acbf-e2f877c85b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_elementDens(cytoband_df, element_type = \"CDS\"):\n",
    "    chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_dict = transcripts_NoOverlaps(element_type)\n",
    "    element_noOverlaps_df = transcripts_dict_toDF(element_dict)\n",
    "    transcript_matrix = element_matrix(chromosmes, element_noOverlaps_df)\n",
    "    colname = element_type+\"s_density\"\n",
    "    calc_elementDensity(cytoband_df, transcript_matrix, colname)\n",
    "    writeNewtxt(cytoPath_write, cytoband_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "68b2db56-016e-4314-a28d-405156aa373a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_intronExonsDens(cytoband_df, element_type = \"CDS\"):\n",
    "    chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_dict = transcripts_NoOverlaps(element_type)\n",
    "    element_noOverlaps_df = transcripts_dict_toDF(element_dict)\n",
    "    exons_dict = transcripts_NoOverlaps(\"exon\")\n",
    "    exons_noOverlaps_df = transcripts_dict_toDF(exons_dict)\n",
    "    print(\"a fazer a matriz...\")\n",
    "    element_introns_exons_matrix = introns_exons_matrix(chromosmes, element_noOverlaps_df, exons_noOverlaps_df)\n",
    "    colname1 = element_type+\"Intron_density\"\n",
    "    colname2 = element_type+\"Exon_density\"\n",
    "    print(\"a fazer os calculos...\")\n",
    "    calc_ExonIntronDensity(cytoband_df, element_introns_exons_matrix, colname1, colname2)\n",
    "    writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9495e9-9702-4571-a78f-9d4835b82301",
   "metadata": {},
   "source": [
    "## CDS density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b2e61635-55f7-4429-9bca-4478b22c7104",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'writeNewtxt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[31], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mstudy_elementDens\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcytobands_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mCDS\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[24], line 8\u001b[0m, in \u001b[0;36mstudy_elementDens\u001b[0;34m(cytoband_df, element_type)\u001b[0m\n\u001b[1;32m      6\u001b[0m colname \u001b[38;5;241m=\u001b[39m element_type\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124ms_density\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      7\u001b[0m calc_elementDensity(cytoband_df, transcript_matrix, colname)\n\u001b[0;32m----> 8\u001b[0m \u001b[43mwriteNewtxt\u001b[49m(cytoPath_write, cytoband_df)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'writeNewtxt' is not defined"
     ]
    }
   ],
   "source": [
    "study_elementDens(cytobands_df, \"CDS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "72d11ba8-f224-4a43-98e3-d7afba89fbbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.016467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.006066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.010266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.005996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.009799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.000917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000050</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  \n",
       "0          0.088       0.326      0.016467  \n",
       "1          0.084       0.282      0.006066  \n",
       "2          0.056       0.275      0.010266  \n",
       "3          0.049       0.405      0.005996  \n",
       "4          0.056       0.404      0.009799  \n",
       "..           ...         ...           ...  \n",
       "858        0.037       0.015      0.000917  \n",
       "859        0.062       0.013      0.002263  \n",
       "860        0.082       0.006      0.001096  \n",
       "861        0.026       0.002      0.001369  \n",
       "862        0.006       0.001      0.000050  \n",
       "\n",
       "[863 rows x 9 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3c8f619e-9ac7-4170-9ec3-b2b1a2303fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CDSs_matrix = element_matrix(chromosmes, CDSs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d65c4c8c-7830-4ae4-928f-2104138ddc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df, CDSs_matrix, \"CDSs_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b4e30cb7-2f45-4a74-8175-fac4802502fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d690d7f-d2aa-453e-b132-1ac6e85dcf68",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within CDSs (Coding Sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3d2701f6-bff2-4b7e-ab92-d701c1bed997",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "study_intronExonsDens(cytobands_df, \"CDS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9b6d07a5-655d-44d4-a2ad-8ea1b2f48ad1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.016467</td>\n",
       "      <td>0.020383</td>\n",
       "      <td>0.013959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>0.010302</td>\n",
       "      <td>0.002620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.010266</td>\n",
       "      <td>0.010090</td>\n",
       "      <td>0.005856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.005996</td>\n",
       "      <td>0.003048</td>\n",
       "      <td>0.001599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.009799</td>\n",
       "      <td>0.010669</td>\n",
       "      <td>0.004489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.000607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.004221</td>\n",
       "      <td>0.002790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>0.002814</td>\n",
       "      <td>0.000916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.001053</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  CDSExon_density  \n",
       "0          0.088       0.326      0.016467           0.020383         0.013959  \n",
       "1          0.084       0.282      0.006066           0.010302         0.002620  \n",
       "2          0.056       0.275      0.010266           0.010090         0.005856  \n",
       "3          0.049       0.405      0.005996           0.003048         0.001599  \n",
       "4          0.056       0.404      0.009799           0.010669         0.004489  \n",
       "..           ...         ...           ...                ...              ...  \n",
       "858        0.037       0.015      0.000917           0.001011         0.000607  \n",
       "859        0.062       0.013      0.002263           0.004221         0.002790  \n",
       "860        0.082       0.006      0.001096           0.002814         0.000916  \n",
       "861        0.026       0.002      0.001369           0.001053         0.000025  \n",
       "862        0.006       0.001      0.000050           0.000082         0.000034  \n",
       "\n",
       "[863 rows x 11 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f257b5cd-3b3b-4d28-8f80-2000a6263843",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CDS_introns_exons_matrix = introns_exons_matrix(chromosmes, CDSs, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6303e000-b492-488a-9f78-3b43abd095a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df, CDS_introns_exons_matrix, \"CDSIntron_density\", \"CDSExon_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "55598914-f945-4356-8645-7eda2944190d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b149728d-101f-4b84-9fa8-a77b5686e219",
   "metadata": {},
   "source": [
    "## Prime 5 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d4d6dc2d-563f-4088-88c3-d623f062ff1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_matrix = element_matrix(chromosmes, UTR5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9737e5af-4410-42d5-a49f-2a19b76d6e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df, UTR5_matrix, \"5P_UTR_density\")\n",
    "#cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2778db1a-3cd0-46d2-961c-4a3b807334a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bd2ce3-76ac-481a-a4af-20733256d243",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within 5 Prime UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "760c6bd2-e12c-43f9-b93b-8ab3980437b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_introns_exons_matrix = introns_exons_matrix(chromosmes, UTR5, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2a1e58ff-3d9e-4b90-b0cc-217029d973af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df2, UTR5_introns_exons_matrix, \"UTR5Intron_density\", \"UTR5Exon_density\")\n",
    "#cytobands_df\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e3aad2b6-cdde-4385-91e8-db90f76dd848",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51f8d5e-d87f-47bc-89ef-60478ed00610",
   "metadata": {},
   "source": [
    "## Prime 3 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2b70b50a-9eae-4fa6-932b-9ce572f7fd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR3_matrix = element_matrix(chromosmes, UTR3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a748357c-732c-4307-81fe-c5e28054425a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_elementDensity(cytobands_df2, UTR3_matrix, \"3P_UTR_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "06edbc35-0c00-46c2-90e2-b9b14eeacd4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d99a6f99-a946-4123-8f99-8984c9b011a0",
   "metadata": {},
   "source": [
    "## Exon and Intron densities within 3 Prime UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a6f4ede0-3475-415d-995e-5096c30268f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR3_introns_exons_matrix = introns_exons_matrix(chromosmes, UTR3, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "33f5773d-6a69-4e55-a936-f88ee3999716",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc_ExonIntronDensity(cytobands_df2, UTR3_introns_exons_matrix, \"UTR3Intron_density\", \"UTR3Exon_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "383b40f1-7696-41e8-bb07-3dceca86bd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc99fd3-b968-4579-9ce9-c0ceec7f6925",
   "metadata": {},
   "source": [
    "## Exons density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b8e17263-abd9-4634-b815-4ef21afd7bf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "#exons_matrix = element_matrix(chromosmes, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b8764313-5f77-4adc-98d1-50d81e0ed8b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \n",
       "0             0.012               0.025             0.019          0.171  \n",
       "1             0.004               0.013             0.006          0.079  \n",
       "2             0.006               0.012             0.010          0.082  \n",
       "3             0.005               0.004             0.004          0.047  \n",
       "4             0.005               0.014             0.010          0.093  \n",
       "..              ...                 ...               ...            ...  \n",
       "858           0.001               0.001             0.003          0.016  \n",
       "859           0.000               0.004             0.002          0.045  \n",
       "860           0.001               0.005             0.002          0.036  \n",
       "861           0.001               0.003             0.001          0.034  \n",
       "862           0.000               0.000             0.000          0.001  \n",
       "\n",
       "[863 rows x 18 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calc_elementDensity(cytobands_df2, exons_matrix, \"Exons_density\")\n",
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6bbc1ed2-3117-4308-b855-aae88c549928",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9d2061bc-6f75-4a28-8156-baf3060b9db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df.values[cytobands_df.index[70]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc6cd1d0-454c-4354-9c18-bd9354568385",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sumOf_columns(df, colums=[\"CDSIntron_density\", \"UTR5Intron_density\", \"UTR3Intron_density\"], newColName = \"Introns_density\"):\n",
    "    \"\"\"\n",
    "    colums is a list of string\n",
    "    \"\"\"\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    new_densites = []\n",
    "    for row in df.index:\n",
    "        density =0\n",
    "        for column in colums:\n",
    "            density += df[column][row]\n",
    "        new_densites.append(round(density, 3))\n",
    "        \n",
    "    df.insert(len(df.columns), newColName, new_densites, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c3c720a7-b910-442a-9125-3bfb4205f7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "      <th>Introns_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "      <td>0.069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \\\n",
       "0             0.012               0.025             0.019          0.171   \n",
       "1             0.004               0.013             0.006          0.079   \n",
       "2             0.006               0.012             0.010          0.082   \n",
       "3             0.005               0.004             0.004          0.047   \n",
       "4             0.005               0.014             0.010          0.093   \n",
       "..              ...                 ...               ...            ...   \n",
       "858           0.001               0.001             0.003          0.016   \n",
       "859           0.000               0.004             0.002          0.045   \n",
       "860           0.001               0.005             0.002          0.036   \n",
       "861           0.001               0.003             0.001          0.034   \n",
       "862           0.000               0.000             0.000          0.001   \n",
       "\n",
       "     Introns_density  \n",
       "0              0.069  \n",
       "1              0.034  \n",
       "2              0.031  \n",
       "3              0.011  \n",
       "4              0.034  \n",
       "..               ...  \n",
       "858            0.003  \n",
       "859            0.012  \n",
       "860            0.013  \n",
       "861            0.008  \n",
       "862            0.000  \n",
       "\n",
       "[863 rows x 19 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sumOf_columns(cytobands_df2)\n",
    "cytobands_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8bf14c9-0b59-42bc-8b73-eec9e7cb7e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586f8eb0-e8fd-4241-8e14-6da5756f8d72",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
