{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a0d93c-1359-4107-807f-f230b3863708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib as mpl\n",
    "import copy\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89dac285-5db9-4261-b4e4-38e0f84db137",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0       1                  2           3            4        5  6  7  \\\n",
       "0        1  GRCh38         chromosome         1.0  248956422.0        .  .  .   \n",
       "1        1       .  biological_region     10469.0      11240.0  1.3e+03  .  .   \n",
       "2        1       .  biological_region     10650.0      10657.0    0.999  +  .   \n",
       "3        1       .  biological_region     10655.0      10657.0    0.999  -  .   \n",
       "4        1       .  biological_region     10678.0      10687.0    0.999  +  .   \n",
       "...     ..     ...                ...         ...          ...      ... .. ..   \n",
       "3516515  Y       .  biological_region  57215804.0   57215813.0    0.999  -  .   \n",
       "3516516  Y       .  biological_region  57215832.0   57215840.0    0.999  +  .   \n",
       "3516517  Y       .  biological_region  57215833.0   57215842.0    0.999  -  .   \n",
       "3516518  Y       .  biological_region  57215862.0   57215867.0    0.999  +  .   \n",
       "3516519  Y       .  biological_region  57215862.0   57215870.0    0.999  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1                 external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2                                       logic_name=eponine  \n",
       "3                                       logic_name=eponine  \n",
       "4                                       logic_name=eponine  \n",
       "...                                                    ...  \n",
       "3516515                                 logic_name=eponine  \n",
       "3516516                                 logic_name=eponine  \n",
       "3516517                                 logic_name=eponine  \n",
       "3516518                                 logic_name=eponine  \n",
       "3516519                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_read= \"/home/androx/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens.GRCh38.112.chr.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f9135a3-08a2-46ed-9c66-2622790d4290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3516515   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3516516   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3516517   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3516518   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3516519   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Unknown1 Unknown2                                        Identifiers  \n",
       "0              .        .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1              .        .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2              +        .                                 logic_name=eponine  \n",
       "3              -        .                                 logic_name=eponine  \n",
       "4              +        .                                 logic_name=eponine  \n",
       "...          ...      ...                                                ...  \n",
       "3516515        -        .                                 logic_name=eponine  \n",
       "3516516        +        .                                 logic_name=eponine  \n",
       "3516517        -        .                                 logic_name=eponine  \n",
       "3516518        +        .                                 logic_name=eponine  \n",
       "3516519        -        .                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Unknown1', 7: 'Unknown2', 8: 'Identifiers'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64551187-3a83-4376-a7db-f0cc87b190b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef8d7245-d6be-4015-bce8-07cb7093450f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328201</th>\n",
       "      <td>10</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133797422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467767</th>\n",
       "      <td>11</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>135086622.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673000</th>\n",
       "      <td>12</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133275309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862634</th>\n",
       "      <td>13</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>114364328.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924501</th>\n",
       "      <td>14</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>107043718.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048326</th>\n",
       "      <td>15</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101991189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171301</th>\n",
       "      <td>16</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90338345.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329042</th>\n",
       "      <td>17</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83257441.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536644</th>\n",
       "      <td>18</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80373285.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598313</th>\n",
       "      <td>19</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58617616.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791801</th>\n",
       "      <td>2</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242193529.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063156</th>\n",
       "      <td>20</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64444167.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145930</th>\n",
       "      <td>21</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46709983.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186784</th>\n",
       "      <td>22</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50818468.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2261287</th>\n",
       "      <td>3</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>198295559.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484781</th>\n",
       "      <td>4</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>190214555.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628956</th>\n",
       "      <td>5</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>181538259.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786252</th>\n",
       "      <td>6</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170805979.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2952608</th>\n",
       "      <td>7</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>159345973.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3120542</th>\n",
       "      <td>8</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145138636.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248106</th>\n",
       "      <td>9</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>138394717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384196</th>\n",
       "      <td>MT</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384360</th>\n",
       "      <td>X</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>156040895.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3505822</th>\n",
       "      <td>Y</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>57227415.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index   Last_index Value Unknown1  \\\n",
       "0         1  GRCh38  chromosome          1.0  248956422.0     .        .   \n",
       "328201   10  GRCh38  chromosome          1.0  133797422.0     .        .   \n",
       "467767   11  GRCh38  chromosome          1.0  135086622.0     .        .   \n",
       "673000   12  GRCh38  chromosome          1.0  133275309.0     .        .   \n",
       "862634   13  GRCh38  chromosome          1.0  114364328.0     .        .   \n",
       "924501   14  GRCh38  chromosome          1.0  107043718.0     .        .   \n",
       "1048326  15  GRCh38  chromosome          1.0  101991189.0     .        .   \n",
       "1171301  16  GRCh38  chromosome          1.0   90338345.0     .        .   \n",
       "1329042  17  GRCh38  chromosome          1.0   83257441.0     .        .   \n",
       "1536644  18  GRCh38  chromosome          1.0   80373285.0     .        .   \n",
       "1598313  19  GRCh38  chromosome          1.0   58617616.0     .        .   \n",
       "1791801   2  GRCh38  chromosome          1.0  242193529.0     .        .   \n",
       "2063156  20  GRCh38  chromosome          1.0   64444167.0     .        .   \n",
       "2145930  21  GRCh38  chromosome          1.0   46709983.0     .        .   \n",
       "2186784  22  GRCh38  chromosome          1.0   50818468.0     .        .   \n",
       "2261287   3  GRCh38  chromosome          1.0  198295559.0     .        .   \n",
       "2484781   4  GRCh38  chromosome          1.0  190214555.0     .        .   \n",
       "2628956   5  GRCh38  chromosome          1.0  181538259.0     .        .   \n",
       "2786252   6  GRCh38  chromosome          1.0  170805979.0     .        .   \n",
       "2952608   7  GRCh38  chromosome          1.0  159345973.0     .        .   \n",
       "3120542   8  GRCh38  chromosome          1.0  145138636.0     .        .   \n",
       "3248106   9  GRCh38  chromosome          1.0  138394717.0     .        .   \n",
       "3384196  MT  GRCh38  chromosome          1.0      16569.0     .        .   \n",
       "3384360   X  GRCh38  chromosome          1.0  156040895.0     .        .   \n",
       "3505822   Y  GRCh38  chromosome          1.0   57227415.0     .        .   \n",
       "\n",
       "        Unknown2                                        Identifiers  \n",
       "0              .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "328201         .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  \n",
       "467767         .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  \n",
       "673000         .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  \n",
       "862634         .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  \n",
       "924501         .  ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...  \n",
       "1048326        .  ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...  \n",
       "1171301        .  ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...  \n",
       "1329042        .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...  \n",
       "1536644        .  ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...  \n",
       "1598313        .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...  \n",
       "1791801        .  ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...  \n",
       "2063156        .  ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...  \n",
       "2145930        .  ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...  \n",
       "2186784        .  ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...  \n",
       "2261287        .  ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...  \n",
       "2484781        .  ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...  \n",
       "2628956        .  ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...  \n",
       "2786252        .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...  \n",
       "2952608        .  ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...  \n",
       "3120542        .  ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...  \n",
       "3248106        .  ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...  \n",
       "3384196        .   ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1  \n",
       "3384360        .  ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...  \n",
       "3505822        .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "chromosmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da5b2aa-931d-4e23-be59-d5b3e2b8e82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2229627f-149a-46f7-9a75-93c56ca4b2b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[\"Identifiers\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "47ddd434-fd68-4235-8d5c-7c69130968c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565.0</td>\n",
       "      <td>65573.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037.0</td>\n",
       "      <td>70008.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000409316;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000329982;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1165</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432.0</td>\n",
       "      <td>924948.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000478421;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516411</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209532.0</td>\n",
       "      <td>57209733.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516413</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209822.0</td>\n",
       "      <td>57209980.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516415</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57210640.0</td>\n",
       "      <td>57210792.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516417</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211552.0</td>\n",
       "      <td>57211620.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516419</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211761.0</td>\n",
       "      <td>57211894.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900145 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome Type  First_index  Last_index Value Unknown1  \\\n",
       "99        1          havana  CDS      65565.0     65573.0     .        +   \n",
       "100       1          havana  CDS      69037.0     70008.0     .        +   \n",
       "398       1  ensembl_havana  CDS     450740.0    451678.0     .        -   \n",
       "544       1  ensembl_havana  CDS     685716.0    686654.0     .        -   \n",
       "1165      1          havana  CDS     924432.0    924948.0     .        +   \n",
       "...      ..             ...  ...          ...         ...   ...      ...   \n",
       "3516411   Y          havana  CDS   57209532.0  57209733.0     .        +   \n",
       "3516413   Y          havana  CDS   57209822.0  57209980.0     .        +   \n",
       "3516415   Y          havana  CDS   57210640.0  57210792.0     .        +   \n",
       "3516417   Y          havana  CDS   57211552.0  57211620.0     .        +   \n",
       "3516419   Y          havana  CDS   57211761.0  57211894.0     .        +   \n",
       "\n",
       "        Unknown2                                        Identifiers  \n",
       "99             0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "100            0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "398            0  ID=CDS:ENSP00000409316;Parent=transcript:ENST0...  \n",
       "544            0  ID=CDS:ENSP00000329982;Parent=transcript:ENST0...  \n",
       "1165           0  ID=CDS:ENSP00000478421;Parent=transcript:ENST0...  \n",
       "...          ...                                                ...  \n",
       "3516411        0  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516413        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516415        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516417        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516419        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "\n",
       "[900145 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "CDSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8af64bb5-1b2d-4493-80a0-939582f088b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516493</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212184.0</td>\n",
       "      <td>57213125.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516494</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213204.0</td>\n",
       "      <td>57213357.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516495</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213526.0</td>\n",
       "      <td>57213602.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516496</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213880.0</td>\n",
       "      <td>57213964.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516497</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214350.0</td>\n",
       "      <td>57214397.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1668627 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome  Type  First_index  Last_index Value Unknown1 Unknown2  \\\n",
       "18        1  havana  exon      11869.0     12227.0     .        +        .   \n",
       "19        1  havana  exon      12613.0     12721.0     .        +        .   \n",
       "20        1  havana  exon      13221.0     14409.0     .        +        .   \n",
       "24        1  havana  exon      12010.0     12057.0     .        +        .   \n",
       "25        1  havana  exon      12179.0     12227.0     .        +        .   \n",
       "...      ..     ...   ...          ...         ...   ...      ...      ...   \n",
       "3516493   Y  havana  exon   57212184.0  57213125.0     .        -        .   \n",
       "3516494   Y  havana  exon   57213204.0  57213357.0     .        -        .   \n",
       "3516495   Y  havana  exon   57213526.0  57213602.0     .        -        .   \n",
       "3516496   Y  havana  exon   57213880.0  57213964.0     .        -        .   \n",
       "3516497   Y  havana  exon   57214350.0  57214397.0     .        -        .   \n",
       "\n",
       "                                               Identifiers  \n",
       "18       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "19       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "20       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "24       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "25       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "...                                                    ...  \n",
       "3516493  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516494  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516495  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516496  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516497  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "\n",
       "[1668627 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ed7c774-ed75-4a23-9548-47d0a9913bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = tanscripts_df.loc[tanscripts_df['Type'] == \"biological_region\"]\n",
    "regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b61a75d-2316-4377-9a54-a7ad6b86b487",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9de20f-329e-4d77-b238-b6f5b84eb879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nao é util para já agregar todos od conjuntos, mais apropriado fazer um mapa de cada coisa\n",
    "\n",
    "def calc_exons_introns(chromosomes_df, sequence_df, exons_df, genes_df, regions_df):\n",
    "    \"\"\"interval_df = Genes or CDSs\n",
    "    +2⁰ = 1: is exon\n",
    "    +2¹ = 2: is part of a region\n",
    "    +2² = 4: is part of CDS\n",
    "    +2³ = 8: is part of gene\n",
    "    \"\"\"\n",
    "    Chrmosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chromosome_elements = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chromosome_exons = exons_df.loc[exons_df['Chr'] == chromosome]\n",
    "        chr_genes = genes_df.loc[genes_df['Chr'] == chromosome]\n",
    "        chr_regions = regions_df.loc[regions_df['Chr'] == chromosome]\n",
    "        element_introns = np.zeros((chr_size), dtype=int)\n",
    "        #genes region\n",
    "        for row in chr_genes.index:\n",
    "            for i in range (int(chr_genes[\"First_index\"][row]), int(chr_genes[\"Last_index\"][row])):\n",
    "                element_introns[i]+=8\n",
    "        #CDS region\n",
    "        for row in chromosome_elements.index:\n",
    "            for i in range (int(chromosome_elements[\"First_index\"][row]), int(chromosome_elements[\"Last_index\"][row])):\n",
    "                element_introns[i]+=4\n",
    "        #other regions\n",
    "        for row in chr_regions.index:\n",
    "            for i in range (int(chr_regions[\"First_index\"][row]), int(chr_regions[\"Last_index\"][row])):\n",
    "                element_introns[i]+=2\n",
    "        #exons region\n",
    "        for row in chromosome_exons.index:\n",
    "            for i in range (int(chromosome_exons[\"First_index\"][row]), int(chromosome_exons[\"Last_index\"][row])):\n",
    "                element_introns[i]+=1\n",
    "        \n",
    "        Chrmosomes_matrix.append(element_introns)\n",
    "    return Chrmosomes_matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "867e6833-f85a-473a-add0-e59b583c5f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_matrix(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    sequence_df = Genes or CDSs, exons or other_regions\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            for i in range (int(chr_sequence[\"First_index\"][row]), int(chr_sequence[\"Last_index\"][row])):\n",
    "                chr_array[i]+=1\n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "741ead14-c552-448e-96f7-2fd45c75f9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def introns_exons_matrix(chromosomes_df, CDS_df, exons_df):\n",
    "    \"\"\"\n",
    "    sequence_df = Genes or CDSs, exons or other_regions\n",
    "    if CDS element: +2\n",
    "    if exon: +1\n",
    "    introns = 2\n",
    "    exons = 3\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_CDS = CDS_df.loc[CDS_df['Chr'] == chromosome]\n",
    "        chromosome_exons = exons_df.loc[exons_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        #CDS region\n",
    "        for row in chr_CDS.index:\n",
    "            for i in range (int(chr_CDS[\"First_index\"][row]), int(chr_CDS[\"Last_index\"][row])):\n",
    "                chr_array[i]+=2\n",
    "        #exons region\n",
    "        for row in chromosome_exons.index:\n",
    "            for i in range (int(chromosome_exons[\"First_index\"][row]), int(chromosome_exons[\"Last_index\"][row])):\n",
    "                chr_array[i]+=1\n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e17263-abd9-4634-b815-4ef21afd7bf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n"
     ]
    }
   ],
   "source": [
    "exons_matrix = element_matrix(chromosmes, exons)\n",
    "exons_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f257b5cd-3b3b-4d28-8f80-2000a6263843",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_matrix = introns_exons_matrix(chromosmes, CDSs, exons)\n",
    "introns_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3184b344-84d9-4278-bccc-7c8807877272",
   "metadata": {},
   "outputs": [],
   "source": [
    "#otherRegions_matrix = calc_group(chromosmes, regions)\n",
    "#otherRegions_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0053bc01-6bdd-46e1-9592-41e0f5a2599b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity(cytobands_df, element_matrix, col_name=\"Element_density\"):\n",
    "    densities = []\n",
    "    chrom_n = 0\n",
    "    for chrom in cytobands_df[\"Chromossome\"].unique():             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Type'] == chrom]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for cytoband in range(len(chr_cytobands.values)):\n",
    "            cytoband_array = element_matrix[chrom_n][int(chr_cytobands[\"First_index\"][cytoband]):int(chr_cytobands[\"Last_index\"][cytoband])]\n",
    "            dr_length = np.count_nonzero(cytoband_array)\n",
    "            cytoband_length = cytobands_chr[\"Size\"][cytoband]\n",
    "            density = dr_length/cytoband_length\n",
    "            densities.append(round(density, 3))\n",
    "            chrom_n +=1\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fd457d08-3d9c-4ad7-b34a-e4cdd5aa389d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  \n",
       "0          0.088       0.326  \n",
       "1          0.084       0.282  \n",
       "2          0.056       0.275  \n",
       "3          0.049       0.405  \n",
       "4          0.056       0.404  \n",
       "..           ...         ...  \n",
       "858        0.037       0.015  \n",
       "859        0.062       0.013  \n",
       "860        0.082       0.006  \n",
       "861        0.026       0.002  \n",
       "862        0.006       0.001  \n",
       "\n",
       "[863 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoPath_read = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoPath_write = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr_dg.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoPath_read, sep=\"\\t\")#, header=None))\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b8764313-5f77-4adc-98d1-50d81e0ed8b3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'exons_matrix' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m calc_elementDensity(cytobands_df, \u001b[43mexons_matrix\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExons_density\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m cytobands_df\n",
      "\u001b[0;31mNameError\u001b[0m: name 'exons_matrix' is not defined"
     ]
    }
   ],
   "source": [
    "calc_elementDensity(cytobands_df, exons_matrix, \"Exons_density\")\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d65c4c8c-7830-4ae4-928f-2104138ddc27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
