{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e14cada1-8d57-4a60-aee0-2cde7c05747f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "920a3ade-d680-4c8f-b532-c645d3416d40",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    str_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        str_cols += col + \"\\t\"\n",
    "    str_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(str_cols))\n",
    "    for row in df.values:\n",
    "        str_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            str_row += str(col) + \"\\t\"\n",
    "        str_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(str_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "19d7cf5b-2e9a-4a6c-9290-c29a965c2949",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gvs_chr20_read= \"/home/androx/Documents/trabalho/datasets/Genomic_variants/chr20/allChr20.GRCH38.sorted.vcf\"\n",
    "#gvs_chr20_df = pd.read_csv(gvs_chr20_read, sep=\"\\t\", usecols=['#CHROM',  'POS', 'REF', 'ALT', 'INFO'])\n",
    "\n",
    "#gvs_chr20_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23fcf1f7-e511-4901-aeef-8da25a818d9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writeNewtxt(gvs_chr20_read, gvs_chr20_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e7e86b23-e2f8-42c3-95ca-8ba253fb183a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gvs_chr20_df[\"INFO\"][0]\n",
    "#gvs_chr20_df[\"INFO\"][30]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f7b849-b735-4670-922d-1223c5dd4e49",
   "metadata": {},
   "source": [
    "### Dividir as genomic variants em SNPs e INDELs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ec3a76c7-e6ab-4b37-85ae-bb0a43e6619f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#snps_chr20_df = get_df_snp(gvs_chr20_df)\n",
    "#snps_chr20_df\n",
    "#mask1 = (gvs_chr20_df['REF'].str.len() == 1) & (gvs_chr20_df['ALT'].str.len() == 1)\n",
    "#snps_chr20_df = gvs_chr20_df.loc[mask1]\n",
    "#snps_chr20_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "864c26a4-ece9-4b0f-bed4-22b81e920377",
   "metadata": {},
   "outputs": [],
   "source": [
    "#indels_chr20_df = get_df_indels(gvs_chr20_df)\n",
    "#indels_chr20_df\n",
    "#mask2 = (gvs_chr20_df['REF'].str.len() > 1) | (gvs_chr20_df['ALT'].str.len() > 1)\n",
    "#indels_chr20_df = gvs_chr20_df.loc[mask2]\n",
    "#indels_chr20_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a912989-b161-4483-8b71-b1de8a442bac",
   "metadata": {},
   "source": [
    "## Seperated density values for insertions and deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "93610664-6a8a-4645-bf8c-15c623dee116",
   "metadata": {},
   "outputs": [],
   "source": [
    "#inserts_chr20 = indels_chr20_df.loc[(indels_chr20_df['REF']) < (indels_chr20_df['ALT'])]\n",
    "#inserts_chr20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dc102cef-513d-42b9-884b-feaf597d26de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#deletions_chr20 = indels_chr20_df.loc[(indels_chr20_df['REF']) > (indels_chr20_df['ALT'])]\n",
    "#deletions_chr20"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ae7cb53-8965-440d-8302-05061860afcc",
   "metadata": {},
   "source": [
    "## Funções de calculo de densidades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4ba47e10-e096-4a02-813f-8723f9b1d120",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_snpDensity(cb_df_chr, gv_df_chr):#, chrom = \"chr20\"):\n",
    "    \"\"\"calculates the density of snps within the cytobands of one chromosome\n",
    "    \"\"\"\n",
    "    densities = []\n",
    "    #chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == chrom]\n",
    "    chr_size = cb_df_chr[\"Last_index\"][cb_df_chr.index[-1]]+1\n",
    "    chr_gvs = np.zeros((chr_size), dtype=int)\n",
    "    #percorrer o dataset pelas vulnerabilidades no cromossoma e anotar na lista np.zeroes\n",
    "    for row in gv_df_chr.index:\n",
    "        chr_gvs[gv_df_chr[\"POS\"][row]]+=1\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for cytoband in cb_df_chr.index:\n",
    "        cytoband_array = chr_gvs[int(cb_df_chr[\"First_index\"][cytoband]):int(cb_df_chr[\"Last_index\"][cytoband])+1]\n",
    "        snp_indel_length = np.count_nonzero(cytoband_array)\n",
    "        cytoband_length = cb_df_chr[\"Size\"][cytoband]\n",
    "        density = snp_indel_length/cytoband_length\n",
    "        densities.append(round(density, 3))\n",
    "        \n",
    "    cb_df_chr.insert(len(cb_df_chr.columns), \"SNP_density\", densities, True)\n",
    "    return cb_df_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "615c744b-8921-408d-9168-f20a06d6f198",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_indelDens(cb_df_chr, gv_df_chr):#, chrom = \"chr20\"):\n",
    "    \"\"\"calculates the density of indels, inserts or deletes within the cytobands of one chromosome\n",
    "    \"\"\"\n",
    "    densities = []\n",
    "    #chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == chrom]\n",
    "    chr_size = cb_df_chr[\"Last_index\"][cb_df_chr.index[-1]]+1\n",
    "    chr_gvs = np.zeros((chr_size), dtype=int)\n",
    "    #percorrer o dataset pelas vulnerabilidades no cromossoma e anotar na lista np.zeroes\n",
    "    for row in gv_df_chr.index:\n",
    "        if len(gv_df_chr[\"REF\"][row]) > len(gv_df_chr[\"ALT\"][row]):               #Deletion\n",
    "            chr_gvs[gv_df_chr[\"POS\"][row]:gv_df_chr[\"POS\"][row] + len(gv_df_chr[\"REF\"][row])+1]+=1\n",
    "        else:                                                                    #Insertion\n",
    "            chr_gvs[gv_df_chr[\"POS\"][row]:gv_df_chr[\"POS\"][row] + len(gv_df_chr[\"ALT\"][row])+1]+=1\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for cytoband in cb_df_chr.index:\n",
    "        cytoband_array = chr_gvs[int(cb_df_chr[\"First_index\"][cytoband]):int(cb_df_chr[\"Last_index\"][cytoband])+1]\n",
    "        snp_indel_length = np.count_nonzero(cytoband_array)\n",
    "        cytoband_length = cb_df_chr[\"Size\"][cytoband]\n",
    "        density = snp_indel_length/cytoband_length\n",
    "        densities.append(round(density, 3))\n",
    "        \n",
    "    cb_df_chr.insert(len(cb_df_chr.columns), \"Indel_density\", densities, True)\n",
    "    return cb_df_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "44f7277d-093f-44f4-b2fc-c2af59b685e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_insert_deleteDens(cb_df_chr, gv_df_chr, dens_type = \"Insertion_density\"):#, chrom = \"chr20\"):\n",
    "    \"\"\"calculates the density of inserts or deletes within the cytobands of one chromosome\n",
    "    \"\"\"\n",
    "    densities = []\n",
    "    #chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == chrom]\n",
    "    chr_size = cb_df_chr[\"Last_index\"][cb_df_chr.index[-1]]+1\n",
    "    chr_gvs = np.zeros((chr_size), dtype=int)\n",
    "    #percorrer o dataset pelas vulnerabilidades no cromossoma e anotar na lista np.zeroes\n",
    "    for row in gv_df_chr.index:\n",
    "        if dens_type == \"Inserts_density\":\n",
    "            chr_gvs[gv_df_chr[\"POS\"][row]:gv_df_chr[\"POS\"][row] + len(gv_df_chr[\"ALT\"][row])+1]+=1\n",
    "        elif dens_type == \"Deletes_density\":\n",
    "            chr_gvs[gv_df_chr[\"POS\"][row]:gv_df_chr[\"POS\"][row] + len(gv_df_chr[\"REF\"][row])+1]+=1\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for cytoband in cb_df_chr.index:\n",
    "        cytoband_array = chr_gvs[int(cb_df_chr[\"First_index\"][cytoband]):int(cb_df_chr[\"Last_index\"][cytoband])+1]\n",
    "        snp_indel_length = np.count_nonzero(cytoband_array)\n",
    "        cytoband_length = cb_df_chr[\"Size\"][cytoband]\n",
    "        density = snp_indel_length/cytoband_length\n",
    "        densities.append(round(density, 3))\n",
    "        \n",
    "    cb_df_chr.insert(len(cb_df_chr.columns), dens_type, densities, True)\n",
    "    return cb_df_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c5683c10-4933-45a7-94a3-f678bb2d2a1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>STR_Density</th>\n",
       "      <th>DR_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>5P_UTR_density</th>\n",
       "      <th>UTR5Intron_density</th>\n",
       "      <th>UTR5Exon_density</th>\n",
       "      <th>3P_UTR_density</th>\n",
       "      <th>UTR3Intron_density</th>\n",
       "      <th>UTR3Exon_density</th>\n",
       "      <th>Exons_density</th>\n",
       "      <th>Introns_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.171</td>\n",
       "      <td>0.069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.282</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.405</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     STR_Density  DR_density  CDSs_density  CDSIntron_density  \\\n",
       "0          0.088       0.326         0.038              0.022   \n",
       "1          0.084       0.282         0.016              0.011   \n",
       "2          0.056       0.275         0.024              0.010   \n",
       "3          0.049       0.405         0.013              0.004   \n",
       "4          0.056       0.404         0.030              0.010   \n",
       "..           ...         ...           ...                ...   \n",
       "858        0.037       0.015         0.004              0.001   \n",
       "859        0.062       0.013         0.002              0.004   \n",
       "860        0.082       0.006         0.005              0.004   \n",
       "861        0.026       0.002         0.003              0.002   \n",
       "862        0.006       0.001         0.000              0.000   \n",
       "\n",
       "     CDSExon_density  5P_UTR_density  UTR5Intron_density  UTR5Exon_density  \\\n",
       "0              0.015           0.033               0.022             0.021   \n",
       "1              0.005           0.022               0.010             0.014   \n",
       "2              0.008           0.034               0.009             0.020   \n",
       "3              0.002           0.018               0.003             0.010   \n",
       "4              0.008           0.036               0.010             0.019   \n",
       "..               ...             ...                 ...               ...   \n",
       "858            0.001           0.003               0.001             0.003   \n",
       "859            0.004           0.001               0.004             0.003   \n",
       "860            0.002           0.004               0.004             0.002   \n",
       "861            0.000           0.002               0.003             0.001   \n",
       "862            0.000           0.000               0.000             0.000   \n",
       "\n",
       "     3P_UTR_density  UTR3Intron_density  UTR3Exon_density  Exons_density  \\\n",
       "0             0.012               0.025             0.019          0.171   \n",
       "1             0.004               0.013             0.006          0.079   \n",
       "2             0.006               0.012             0.010          0.082   \n",
       "3             0.005               0.004             0.004          0.047   \n",
       "4             0.005               0.014             0.010          0.093   \n",
       "..              ...                 ...               ...            ...   \n",
       "858           0.001               0.001             0.003          0.016   \n",
       "859           0.000               0.004             0.002          0.045   \n",
       "860           0.001               0.005             0.002          0.036   \n",
       "861           0.001               0.003             0.001          0.034   \n",
       "862           0.000               0.000             0.000          0.001   \n",
       "\n",
       "     Introns_density  \n",
       "0              0.069  \n",
       "1              0.034  \n",
       "2              0.031  \n",
       "3              0.011  \n",
       "4              0.034  \n",
       "..               ...  \n",
       "858            0.003  \n",
       "859            0.012  \n",
       "860            0.013  \n",
       "861            0.008  \n",
       "862            0.000  \n",
       "\n",
       "[863 rows x 19 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoPath_read = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr_dg.txt\"\n",
    "cytoPath_write = \"/home/androx/Documents/trabalho/citobands/cytobandFiltered_processed_str_dr_dg_gv.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoPath_read, sep=\"\\t\")#, header=None))\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "049abdc9-f901-4cb4-80c6-a1aeea101af2",
   "metadata": {},
   "source": [
    "### Cytoband data frame only chr20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "641efa15-37fa-4de3-8e87-7af09f1c4145",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df = cytobands_df.loc[cytobands_df['Chromossome'] == \"chr20\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3550cf04-69b7-4925-822d-59d02f87a02b",
   "metadata": {},
   "source": [
    "### Calculate the different densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c50e1535-fa8b-4b0e-a109-5ba0b625886b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df = calc_snpDensity(chr20_df, snps_chr20_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2f15865e-2dcb-4a4a-9bfd-4b6314b3cdd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df = calc_indelDens(chr20_df, indels_chr20_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5bfa5183-6001-4ef9-aca0-03276322ad7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df = calc_insert_deleteDens(chr20_df, inserts_chr20, \"Insertion_density\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "43fe709b-7668-4d9a-b5e9-0f4e429b3e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df = calc_insert_deleteDens(chr20_df, deletions_chr20, \"Deletion_density\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "596acb7f-db0f-4f12-bd73-c9683c55cd0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr20_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b4d09ff-6e34-45d0-8ac0-223ac3e0734a",
   "metadata": {},
   "source": [
    "### Make plots of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8aa4a108-8bcd-42cf-8fcd-80012e890121",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_1chr(chr_df, chromossome, col = \"SNP_density\"):\n",
    "    plt.figure(figsize=(30,10))\n",
    "    plt.plot(chr_df[\"Citoband\"], chr_df[col], 'ro--', linewidth=2, markersize=6)\n",
    "    plt.xlabel(\"Cytoband\")\n",
    "    plt.ylabel(col)\n",
    "    plt.grid()\n",
    "    plt.title(col + \" in chromosome \"  + chromossome)\n",
    "    #plt.show()\n",
    "    plt.savefig(chromossome + col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61450f1b-f18a-470e-9480-08e4da8b8301",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_1chr(chr20_df, \"20\", \"SNP_density\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "896d30ee-94ea-4781-8d89-0adf046b6fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_1chr(chr20_df, \"20\", \"Indel_density\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6ddde718-3933-46fd-bff8-0a18c664cb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytoPath_write = \"/home/androx/Documents/trabalho/citobands/Latest_gvs/cytobandFiltered_processed_str_dr_dg_gv_chr20.txt\"\n",
    "#writeNewtxt(cytoPath_write, chr20_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4baa9745-8668-4468-ad7d-9cc38a55e389",
   "metadata": {},
   "source": [
    "# Codigo de execução para cada um dos chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "94ca039d-747e-48ee-bf5c-703be05afece",
   "metadata": {},
   "outputs": [],
   "source": [
    "def execute_1chr(chromosome):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    gvs_chr_read= \"/home/androx/Documents/trabalho/datasets/Genomic_variants/\" + chromosome + \"/\" + chromosome + \"_gvs.vcf\"\n",
    "    cytoband_write = \"/home/androx/Documents/trabalho/citobands/Latest_gvs/cytobandFiltered_processed_str_dr_dg_gv_\" + chromosome + \".txt\"\n",
    "    \n",
    "    print(gvs_chr_read)\n",
    "    gvsChr_df = pd.read_csv(gvs_chr_read, sep=\"\\t\")           #genomic variants dataset\n",
    "\n",
    "    ####### Separation of gv dataset into snps, indels, inserts and delets datasets ############\n",
    "    mask1 = (gvsChr_df['REF'].str.len() == 1) & (gvsChr_df['ALT'].str.len() == 1)\n",
    "    mask2 = (gvsChr_df['REF'].str.len() > 1) | (gvsChr_df['ALT'].str.len() > 1)\n",
    "    \n",
    "    snps_chr_df = gvsChr_df.loc[mask1]\n",
    "    indels_chr_df = gvsChr_df.loc[mask2]\n",
    "    inserts_chr_df = indels_chr_df.loc[(indels_chr_df['REF']) < (indels_chr_df['ALT'])]\n",
    "    deletes_chr_df = indels_chr_df.loc[(indels_chr_df['REF']) > (indels_chr_df['ALT'])]\n",
    "\n",
    "    #print(indels_chr_df)\n",
    "    ####### Calculation of densities into the cytoband dataset\n",
    "    chr_df = cytobands_df.loc[cytobands_df['Chromossome'] == chromosome]    #cytobands dataset of intended chromosome\n",
    "    \n",
    "    chr_df = calc_snpDensity(chr_df, snps_chr_df)\n",
    "    chr_df = calc_indelDens(chr_df, indels_chr_df)\n",
    "    chr_df = calc_insert_deleteDens(chr_df, inserts_chr_df, \"Inserts_density\")\n",
    "    chr_df = calc_insert_deleteDens(chr_df, deletes_chr_df, \"Deletes_density\")\n",
    "    print(chr_df)\n",
    "\n",
    "    writeNewtxt(cytoband_write, chr_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4811970c-bcdc-449a-b943-3c8a2885207b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/androx/Documents/trabalho/datasets/Genomic_variants/chr3/chr3_gvs.vcf\n",
      "    Chromossome  First_index  Last_index Citoband  Unkown     Size  \\\n",
      "482        chr3            0     2799999    p26.3  gpos50  2799999   \n",
      "483        chr3      2800000     3999999    p26.2    gneg  1199999   \n",
      "484        chr3      4000000     8099999    p26.1  gpos50  4099999   \n",
      "485        chr3      8100000    11599999    p25.3    gneg  3499999   \n",
      "486        chr3     11600000    13199999    p25.2  gpos25  1599999   \n",
      "..          ...          ...         ...      ...     ...      ...   \n",
      "539        chr3    183000000   184799999    q27.1    gneg  1799999   \n",
      "540        chr3    184800000   186299999    q27.2  gpos25  1499999   \n",
      "541        chr3    186300000   188199999    q27.3    gneg  1899999   \n",
      "542        chr3    188200000   192599999      q28  gpos75  4399999   \n",
      "543        chr3    192600000   198295558      q29    gneg  5695558   \n",
      "\n",
      "     STR_Density  DR_density  CDSs_density  CDSIntron_density  ...  \\\n",
      "482        0.020       0.041         0.003              0.002  ...   \n",
      "483        0.012       0.123         0.008              0.008  ...   \n",
      "484        0.017       0.149         0.005              0.002  ...   \n",
      "485        0.032       0.292         0.020              0.013  ...   \n",
      "486        0.031       0.341         0.013              0.010  ...   \n",
      "..           ...         ...           ...                ...  ...   \n",
      "539        0.036       0.348         0.036              0.017  ...   \n",
      "540        0.032       0.257         0.014              0.004  ...   \n",
      "541        0.023       0.347         0.011              0.008  ...   \n",
      "542        0.019       0.225         0.004              0.004  ...   \n",
      "543        0.057       0.321         0.017              0.012  ...   \n",
      "\n",
      "     UTR5Exon_density  3P_UTR_density  UTR3Intron_density  UTR3Exon_density  \\\n",
      "482             0.001           0.001               0.002             0.002   \n",
      "483             0.006           0.002               0.008             0.002   \n",
      "484             0.004           0.001               0.002             0.002   \n",
      "485             0.015           0.008               0.013             0.008   \n",
      "486             0.005           0.011               0.009             0.013   \n",
      "..                ...             ...                 ...               ...   \n",
      "539             0.014           0.010               0.017             0.014   \n",
      "540             0.023           0.003               0.005             0.005   \n",
      "541             0.009           0.006               0.009             0.010   \n",
      "542             0.007           0.002               0.004             0.002   \n",
      "543             0.015           0.005               0.014             0.007   \n",
      "\n",
      "     Exons_density  Introns_density  SNP_density  Indel_density  \\\n",
      "482          0.020            0.005        0.035          0.010   \n",
      "483          0.041            0.019        0.034          0.009   \n",
      "484          0.025            0.006        0.033          0.009   \n",
      "485          0.090            0.036        0.029          0.008   \n",
      "486          0.065            0.026        0.028          0.009   \n",
      "..             ...              ...          ...            ...   \n",
      "539          0.118            0.043        0.027          0.009   \n",
      "540          0.067            0.012        0.025          0.009   \n",
      "541          0.067            0.025        0.029          0.008   \n",
      "542          0.029            0.010        0.028          0.009   \n",
      "543          0.089            0.035        0.027          0.009   \n",
      "\n",
      "     Inserts_density  Deletes_density  \n",
      "482            0.003            0.007  \n",
      "483            0.003            0.006  \n",
      "484            0.002            0.007  \n",
      "485            0.002            0.006  \n",
      "486            0.002            0.006  \n",
      "..               ...              ...  \n",
      "539            0.002            0.006  \n",
      "540            0.002            0.007  \n",
      "541            0.002            0.006  \n",
      "542            0.002            0.007  \n",
      "543            0.002            0.007  \n",
      "\n",
      "[62 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "execute_1chr(\"chr4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96a0eff-1f77-45ab-9e1f-75c33b178674",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
